{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Untitled85.ipynb",
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    },
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {
        "0595cc45a97d4436a419113d34d82b36": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "state": {
            "_view_name": "HBoxView",
            "_dom_classes": [],
            "_model_name": "HBoxModel",
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "box_style": "",
            "layout": "IPY_MODEL_e8a5f6a41f4244368a1e5cad5bca7b25",
            "_model_module": "@jupyter-widgets/controls",
            "children": [
              "IPY_MODEL_8703c475a615485f8358aec5634854c6",
              "IPY_MODEL_9defc53806e24e57b8ee9188efb2f438"
            ]
          }
        },
        "e8a5f6a41f4244368a1e5cad5bca7b25": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        },
        "8703c475a615485f8358aec5634854c6": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "state": {
            "_view_name": "ProgressView",
            "style": "IPY_MODEL_1dd85dfd78d64ce2bcd4339d250d7d43",
            "_dom_classes": [],
            "description": "Batches: 100%",
            "_model_name": "FloatProgressModel",
            "bar_style": "success",
            "max": 4,
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "value": 4,
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "orientation": "horizontal",
            "min": 0,
            "description_tooltip": null,
            "_model_module": "@jupyter-widgets/controls",
            "layout": "IPY_MODEL_6bb25d52fb19411ba1ff5710bc165a9a"
          }
        },
        "9defc53806e24e57b8ee9188efb2f438": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "state": {
            "_view_name": "HTMLView",
            "style": "IPY_MODEL_9b4397d6c26f4d62808c6772b29a2f11",
            "_dom_classes": [],
            "description": "",
            "_model_name": "HTMLModel",
            "placeholder": "​",
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "value": " 4/4 [00:11&lt;00:00,  2.80s/it]",
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "description_tooltip": null,
            "_model_module": "@jupyter-widgets/controls",
            "layout": "IPY_MODEL_af0a28394d034f4882051a1c945b0272"
          }
        },
        "1dd85dfd78d64ce2bcd4339d250d7d43": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "state": {
            "_view_name": "StyleView",
            "_model_name": "ProgressStyleModel",
            "description_width": "initial",
            "_view_module": "@jupyter-widgets/base",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.2.0",
            "bar_color": null,
            "_model_module": "@jupyter-widgets/controls"
          }
        },
        "6bb25d52fb19411ba1ff5710bc165a9a": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        },
        "9b4397d6c26f4d62808c6772b29a2f11": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_view_name": "StyleView",
            "_model_name": "DescriptionStyleModel",
            "description_width": "",
            "_view_module": "@jupyter-widgets/base",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.2.0",
            "_model_module": "@jupyter-widgets/controls"
          }
        },
        "af0a28394d034f4882051a1c945b0272": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        },
        "87e75f2764fa41908bc106cc5c2a68bf": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "state": {
            "_view_name": "HBoxView",
            "_dom_classes": [],
            "_model_name": "HBoxModel",
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "box_style": "",
            "layout": "IPY_MODEL_a7e610734d6544bfa6f82b0e0f272110",
            "_model_module": "@jupyter-widgets/controls",
            "children": [
              "IPY_MODEL_39b061cb06bd4fe6a2eb675f2987e253",
              "IPY_MODEL_817cb5cc10894bb9943b48d89861c9a5"
            ]
          }
        },
        "a7e610734d6544bfa6f82b0e0f272110": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        },
        "39b061cb06bd4fe6a2eb675f2987e253": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "state": {
            "_view_name": "ProgressView",
            "style": "IPY_MODEL_47d57a057d734cea8f01f5f857bc903e",
            "_dom_classes": [],
            "description": "Batches: 100%",
            "_model_name": "FloatProgressModel",
            "bar_style": "success",
            "max": 4,
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "value": 4,
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "orientation": "horizontal",
            "min": 0,
            "description_tooltip": null,
            "_model_module": "@jupyter-widgets/controls",
            "layout": "IPY_MODEL_c9c8573e5f7c4b63bf8d20707580b81e"
          }
        },
        "817cb5cc10894bb9943b48d89861c9a5": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "state": {
            "_view_name": "HTMLView",
            "style": "IPY_MODEL_99288fe0d15341128db52f71d9536ee3",
            "_dom_classes": [],
            "description": "",
            "_model_name": "HTMLModel",
            "placeholder": "​",
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "value": " 4/4 [00:06&lt;00:00,  1.74s/it]",
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "description_tooltip": null,
            "_model_module": "@jupyter-widgets/controls",
            "layout": "IPY_MODEL_b2b0a8e1cc6d46e18b499009af7230a1"
          }
        },
        "47d57a057d734cea8f01f5f857bc903e": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "state": {
            "_view_name": "StyleView",
            "_model_name": "ProgressStyleModel",
            "description_width": "initial",
            "_view_module": "@jupyter-widgets/base",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.2.0",
            "bar_color": null,
            "_model_module": "@jupyter-widgets/controls"
          }
        },
        "c9c8573e5f7c4b63bf8d20707580b81e": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        },
        "99288fe0d15341128db52f71d9536ee3": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_view_name": "StyleView",
            "_model_name": "DescriptionStyleModel",
            "description_width": "",
            "_view_module": "@jupyter-widgets/base",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.2.0",
            "_model_module": "@jupyter-widgets/controls"
          }
        },
        "b2b0a8e1cc6d46e18b499009af7230a1": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        },
        "4864f597528a455690b3786b61b81a78": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "state": {
            "_view_name": "HBoxView",
            "_dom_classes": [],
            "_model_name": "HBoxModel",
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "box_style": "",
            "layout": "IPY_MODEL_fa8b785a948447178b5695857a9852a1",
            "_model_module": "@jupyter-widgets/controls",
            "children": [
              "IPY_MODEL_c67964d3c9c549f9ba3e5fae610e47f0",
              "IPY_MODEL_cd75a3ae6682496085415b48ed3a7038"
            ]
          }
        },
        "fa8b785a948447178b5695857a9852a1": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        },
        "c67964d3c9c549f9ba3e5fae610e47f0": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "state": {
            "_view_name": "ProgressView",
            "style": "IPY_MODEL_e60cae3ceed54230b6d53c5af2505351",
            "_dom_classes": [],
            "description": "Batches: 100%",
            "_model_name": "FloatProgressModel",
            "bar_style": "success",
            "max": 6,
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "value": 6,
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "orientation": "horizontal",
            "min": 0,
            "description_tooltip": null,
            "_model_module": "@jupyter-widgets/controls",
            "layout": "IPY_MODEL_06065f123a474ef3aae7529522d3ac01"
          }
        },
        "cd75a3ae6682496085415b48ed3a7038": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "state": {
            "_view_name": "HTMLView",
            "style": "IPY_MODEL_a3df8c95d38e413cad50b64ff253252e",
            "_dom_classes": [],
            "description": "",
            "_model_name": "HTMLModel",
            "placeholder": "​",
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "value": " 6/6 [00:04&lt;00:00,  1.45it/s]",
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "description_tooltip": null,
            "_model_module": "@jupyter-widgets/controls",
            "layout": "IPY_MODEL_87323962ea2e4866a82de3e9a3ec0888"
          }
        },
        "e60cae3ceed54230b6d53c5af2505351": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "state": {
            "_view_name": "StyleView",
            "_model_name": "ProgressStyleModel",
            "description_width": "initial",
            "_view_module": "@jupyter-widgets/base",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.2.0",
            "bar_color": null,
            "_model_module": "@jupyter-widgets/controls"
          }
        },
        "06065f123a474ef3aae7529522d3ac01": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        },
        "a3df8c95d38e413cad50b64ff253252e": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_view_name": "StyleView",
            "_model_name": "DescriptionStyleModel",
            "description_width": "",
            "_view_module": "@jupyter-widgets/base",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.2.0",
            "_model_module": "@jupyter-widgets/controls"
          }
        },
        "87323962ea2e4866a82de3e9a3ec0888": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        },
        "0ac253bc93ac4d27b1f29dfe1eeb5332": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "state": {
            "_view_name": "HBoxView",
            "_dom_classes": [],
            "_model_name": "HBoxModel",
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "box_style": "",
            "layout": "IPY_MODEL_45c5d6b64352490db3b4266395d4fa44",
            "_model_module": "@jupyter-widgets/controls",
            "children": [
              "IPY_MODEL_53b970bc82c44adca9289066e4991b56",
              "IPY_MODEL_37bf76bec6a34d0a92a7887eef085a3b"
            ]
          }
        },
        "45c5d6b64352490db3b4266395d4fa44": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        },
        "53b970bc82c44adca9289066e4991b56": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "state": {
            "_view_name": "ProgressView",
            "style": "IPY_MODEL_d4f9540213ad40e5836ffb0e5aeefbc6",
            "_dom_classes": [],
            "description": "Batches: 100%",
            "_model_name": "FloatProgressModel",
            "bar_style": "success",
            "max": 4,
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "value": 4,
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "orientation": "horizontal",
            "min": 0,
            "description_tooltip": null,
            "_model_module": "@jupyter-widgets/controls",
            "layout": "IPY_MODEL_f60fa13ade1e4141a6859871af99539d"
          }
        },
        "37bf76bec6a34d0a92a7887eef085a3b": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "state": {
            "_view_name": "HTMLView",
            "style": "IPY_MODEL_135de97544d34137bcb2278876347a73",
            "_dom_classes": [],
            "description": "",
            "_model_name": "HTMLModel",
            "placeholder": "​",
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "value": " 4/4 [00:03&lt;00:00,  1.09it/s]",
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "description_tooltip": null,
            "_model_module": "@jupyter-widgets/controls",
            "layout": "IPY_MODEL_da3e6e15544a489c9acc0c80a8f12b8c"
          }
        },
        "d4f9540213ad40e5836ffb0e5aeefbc6": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "state": {
            "_view_name": "StyleView",
            "_model_name": "ProgressStyleModel",
            "description_width": "initial",
            "_view_module": "@jupyter-widgets/base",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.2.0",
            "bar_color": null,
            "_model_module": "@jupyter-widgets/controls"
          }
        },
        "f60fa13ade1e4141a6859871af99539d": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        },
        "135de97544d34137bcb2278876347a73": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_view_name": "StyleView",
            "_model_name": "DescriptionStyleModel",
            "description_width": "",
            "_view_module": "@jupyter-widgets/base",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.2.0",
            "_model_module": "@jupyter-widgets/controls"
          }
        },
        "da3e6e15544a489c9acc0c80a8f12b8c": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        }
      }
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nQy7XKtQ3ZHv"
      },
      "source": [
        "# Installs"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "kJYf4CYG2tCG",
        "outputId": "5dcd252d-cdb5-4275-f6e5-c027cb0fa62b"
      },
      "source": [
        "pip install -U sentence-transformers"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting sentence-transformers\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/35/aa/f672ce489063c4ee7a566ebac1b723c53ac0cea19d9e36599cc241d8ed56/sentence-transformers-1.0.4.tar.gz (74kB)\n",
            "\u001b[K     |████████████████████████████████| 81kB 3.2MB/s \n",
            "\u001b[?25hCollecting transformers<5.0.0,>=3.1.0\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/ed/d5/f4157a376b8a79489a76ce6cfe147f4f3be1e029b7144fa7b8432e8acb26/transformers-4.4.2-py3-none-any.whl (2.0MB)\n",
            "\u001b[K     |████████████████████████████████| 2.0MB 6.0MB/s \n",
            "\u001b[?25hRequirement already satisfied, skipping upgrade: tqdm in /usr/local/lib/python3.7/dist-packages (from sentence-transformers) (4.41.1)\n",
            "Requirement already satisfied, skipping upgrade: torch>=1.6.0 in /usr/local/lib/python3.7/dist-packages (from sentence-transformers) (1.8.1+cu101)\n",
            "Requirement already satisfied, skipping upgrade: numpy in /usr/local/lib/python3.7/dist-packages (from sentence-transformers) (1.19.5)\n",
            "Requirement already satisfied, skipping upgrade: scikit-learn in /usr/local/lib/python3.7/dist-packages (from sentence-transformers) (0.22.2.post1)\n",
            "Requirement already satisfied, skipping upgrade: scipy in /usr/local/lib/python3.7/dist-packages (from sentence-transformers) (1.4.1)\n",
            "Requirement already satisfied, skipping upgrade: nltk in /usr/local/lib/python3.7/dist-packages (from sentence-transformers) (3.2.5)\n",
            "Collecting sentencepiece\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/f5/99/e0808cb947ba10f575839c43e8fafc9cc44e4a7a2c8f79c60db48220a577/sentencepiece-0.1.95-cp37-cp37m-manylinux2014_x86_64.whl (1.2MB)\n",
            "\u001b[K     |████████████████████████████████| 1.2MB 19.5MB/s \n",
            "\u001b[?25hRequirement already satisfied, skipping upgrade: filelock in /usr/local/lib/python3.7/dist-packages (from transformers<5.0.0,>=3.1.0->sentence-transformers) (3.0.12)\n",
            "Requirement already satisfied, skipping upgrade: requests in /usr/local/lib/python3.7/dist-packages (from transformers<5.0.0,>=3.1.0->sentence-transformers) (2.23.0)\n",
            "Requirement already satisfied, skipping upgrade: importlib-metadata; python_version < \"3.8\" in /usr/local/lib/python3.7/dist-packages (from transformers<5.0.0,>=3.1.0->sentence-transformers) (3.8.1)\n",
            "Collecting tokenizers<0.11,>=0.10.1\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/71/23/2ddc317b2121117bf34dd00f5b0de194158f2a44ee2bf5e47c7166878a97/tokenizers-0.10.1-cp37-cp37m-manylinux2010_x86_64.whl (3.2MB)\n",
            "\u001b[K     |████████████████████████████████| 3.2MB 28.2MB/s \n",
            "\u001b[?25hCollecting sacremoses\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/7d/34/09d19aff26edcc8eb2a01bed8e98f13a1537005d31e95233fd48216eed10/sacremoses-0.0.43.tar.gz (883kB)\n",
            "\u001b[K     |████████████████████████████████| 890kB 26.0MB/s \n",
            "\u001b[?25hRequirement already satisfied, skipping upgrade: regex!=2019.12.17 in /usr/local/lib/python3.7/dist-packages (from transformers<5.0.0,>=3.1.0->sentence-transformers) (2019.12.20)\n",
            "Requirement already satisfied, skipping upgrade: packaging in /usr/local/lib/python3.7/dist-packages (from transformers<5.0.0,>=3.1.0->sentence-transformers) (20.9)\n",
            "Requirement already satisfied, skipping upgrade: typing-extensions in /usr/local/lib/python3.7/dist-packages (from torch>=1.6.0->sentence-transformers) (3.7.4.3)\n",
            "Requirement already satisfied, skipping upgrade: joblib>=0.11 in /usr/local/lib/python3.7/dist-packages (from scikit-learn->sentence-transformers) (1.0.1)\n",
            "Requirement already satisfied, skipping upgrade: six in /usr/local/lib/python3.7/dist-packages (from nltk->sentence-transformers) (1.15.0)\n",
            "Requirement already satisfied, skipping upgrade: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->transformers<5.0.0,>=3.1.0->sentence-transformers) (2020.12.5)\n",
            "Requirement already satisfied, skipping upgrade: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests->transformers<5.0.0,>=3.1.0->sentence-transformers) (3.0.4)\n",
            "Requirement already satisfied, skipping upgrade: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests->transformers<5.0.0,>=3.1.0->sentence-transformers) (2.10)\n",
            "Requirement already satisfied, skipping upgrade: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->transformers<5.0.0,>=3.1.0->sentence-transformers) (1.24.3)\n",
            "Requirement already satisfied, skipping upgrade: zipp>=0.5 in /usr/local/lib/python3.7/dist-packages (from importlib-metadata; python_version < \"3.8\"->transformers<5.0.0,>=3.1.0->sentence-transformers) (3.4.1)\n",
            "Requirement already satisfied, skipping upgrade: click in /usr/local/lib/python3.7/dist-packages (from sacremoses->transformers<5.0.0,>=3.1.0->sentence-transformers) (7.1.2)\n",
            "Requirement already satisfied, skipping upgrade: pyparsing>=2.0.2 in /usr/local/lib/python3.7/dist-packages (from packaging->transformers<5.0.0,>=3.1.0->sentence-transformers) (2.4.7)\n",
            "Building wheels for collected packages: sentence-transformers, sacremoses\n",
            "  Building wheel for sentence-transformers (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for sentence-transformers: filename=sentence_transformers-1.0.4-cp37-none-any.whl size=114307 sha256=f616fab2215213e5d2710844ddc07f3bbc6910d6d1bbc3c722363c94381619e8\n",
            "  Stored in directory: /root/.cache/pip/wheels/26/ea/89/d0d2e013d951b6d23270aa9ca4018b82632ab7cd933c331316\n",
            "  Building wheel for sacremoses (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for sacremoses: filename=sacremoses-0.0.43-cp37-none-any.whl size=893262 sha256=8932018b136239790bba0b1dfcd3046a7c336c7bb34d2d766f1f06c2247c6d52\n",
            "  Stored in directory: /root/.cache/pip/wheels/29/3c/fd/7ce5c3f0666dab31a50123635e6fb5e19ceb42ce38d4e58f45\n",
            "Successfully built sentence-transformers sacremoses\n",
            "Installing collected packages: tokenizers, sacremoses, transformers, sentencepiece, sentence-transformers\n",
            "Successfully installed sacremoses-0.0.43 sentence-transformers-1.0.4 sentencepiece-0.1.95 tokenizers-0.10.1 transformers-4.4.2\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "cISG5vn-nNhA",
        "outputId": "f6ce4bbe-92f4-4ae1-b680-75910352cbf3"
      },
      "source": [
        "pip install pdfminer.six"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting pdfminer.six\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/93/f3/4fec7dabe8802ebec46141345bf714cd1fc7d93cb74ddde917e4b6d97d88/pdfminer.six-20201018-py3-none-any.whl (5.6MB)\n",
            "\u001b[K     |████████████████████████████████| 5.6MB 5.3MB/s \n",
            "\u001b[?25hRequirement already satisfied: chardet; python_version > \"3.0\" in /usr/local/lib/python3.7/dist-packages (from pdfminer.six) (3.0.4)\n",
            "Collecting cryptography\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/b2/26/7af637e6a7e87258b963f1731c5982fb31cd507f0d90d91836e446955d02/cryptography-3.4.7-cp36-abi3-manylinux2014_x86_64.whl (3.2MB)\n",
            "\u001b[K     |████████████████████████████████| 3.2MB 38.2MB/s \n",
            "\u001b[?25hRequirement already satisfied: sortedcontainers in /usr/local/lib/python3.7/dist-packages (from pdfminer.six) (2.3.0)\n",
            "Requirement already satisfied: cffi>=1.12 in /usr/local/lib/python3.7/dist-packages (from cryptography->pdfminer.six) (1.14.5)\n",
            "Requirement already satisfied: pycparser in /usr/local/lib/python3.7/dist-packages (from cffi>=1.12->cryptography->pdfminer.six) (2.20)\n",
            "Installing collected packages: cryptography, pdfminer.six\n",
            "Successfully installed cryptography-3.4.7 pdfminer.six-20201018\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ek7jbhX52RUT",
        "outputId": "4b19962a-b39d-4021-c8f8-10130249920f"
      },
      "source": [
        "pip install transformers torch"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: transformers in /usr/local/lib/python3.7/dist-packages (4.4.2)\n",
            "Requirement already satisfied: torch in /usr/local/lib/python3.7/dist-packages (1.8.1+cu101)\n",
            "Requirement already satisfied: tqdm>=4.27 in /usr/local/lib/python3.7/dist-packages (from transformers) (4.41.1)\n",
            "Requirement already satisfied: regex!=2019.12.17 in /usr/local/lib/python3.7/dist-packages (from transformers) (2019.12.20)\n",
            "Requirement already satisfied: importlib-metadata; python_version < \"3.8\" in /usr/local/lib/python3.7/dist-packages (from transformers) (3.8.1)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.7/dist-packages (from transformers) (20.9)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.7/dist-packages (from transformers) (3.0.12)\n",
            "Requirement already satisfied: tokenizers<0.11,>=0.10.1 in /usr/local/lib/python3.7/dist-packages (from transformers) (0.10.1)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.7/dist-packages (from transformers) (2.23.0)\n",
            "Requirement already satisfied: numpy>=1.17 in /usr/local/lib/python3.7/dist-packages (from transformers) (1.19.5)\n",
            "Requirement already satisfied: sacremoses in /usr/local/lib/python3.7/dist-packages (from transformers) (0.0.43)\n",
            "Requirement already satisfied: typing-extensions in /usr/local/lib/python3.7/dist-packages (from torch) (3.7.4.3)\n",
            "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.7/dist-packages (from importlib-metadata; python_version < \"3.8\"->transformers) (3.4.1)\n",
            "Requirement already satisfied: pyparsing>=2.0.2 in /usr/local/lib/python3.7/dist-packages (from packaging->transformers) (2.4.7)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->transformers) (1.24.3)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->transformers) (2020.12.5)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests->transformers) (2.10)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests->transformers) (3.0.4)\n",
            "Requirement already satisfied: six in /usr/local/lib/python3.7/dist-packages (from sacremoses->transformers) (1.15.0)\n",
            "Requirement already satisfied: click in /usr/local/lib/python3.7/dist-packages (from sacremoses->transformers) (7.1.2)\n",
            "Requirement already satisfied: joblib in /usr/local/lib/python3.7/dist-packages (from sacremoses->transformers) (1.0.1)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "HzG92jY_2avD",
        "outputId": "1c02e76b-feaf-478a-be7d-123e1932844d"
      },
      "source": [
        "pip install bert-extractive-summarizer"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting bert-extractive-summarizer\n",
            "  Downloading https://files.pythonhosted.org/packages/1a/07/fdb05f9e18b6f641499ef56737126fbd2fafe1cdc1a04ba069d5aa205901/bert_extractive_summarizer-0.7.1-py3-none-any.whl\n",
            "Requirement already satisfied: scikit-learn in /usr/local/lib/python3.7/dist-packages (from bert-extractive-summarizer) (0.22.2.post1)\n",
            "Requirement already satisfied: spacy in /usr/local/lib/python3.7/dist-packages (from bert-extractive-summarizer) (2.2.4)\n",
            "Requirement already satisfied: transformers in /usr/local/lib/python3.7/dist-packages (from bert-extractive-summarizer) (4.4.2)\n",
            "Requirement already satisfied: joblib>=0.11 in /usr/local/lib/python3.7/dist-packages (from scikit-learn->bert-extractive-summarizer) (1.0.1)\n",
            "Requirement already satisfied: scipy>=0.17.0 in /usr/local/lib/python3.7/dist-packages (from scikit-learn->bert-extractive-summarizer) (1.4.1)\n",
            "Requirement already satisfied: numpy>=1.11.0 in /usr/local/lib/python3.7/dist-packages (from scikit-learn->bert-extractive-summarizer) (1.19.5)\n",
            "Requirement already satisfied: tqdm<5.0.0,>=4.38.0 in /usr/local/lib/python3.7/dist-packages (from spacy->bert-extractive-summarizer) (4.41.1)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.7/dist-packages (from spacy->bert-extractive-summarizer) (54.2.0)\n",
            "Requirement already satisfied: wasabi<1.1.0,>=0.4.0 in /usr/local/lib/python3.7/dist-packages (from spacy->bert-extractive-summarizer) (0.8.2)\n",
            "Requirement already satisfied: srsly<1.1.0,>=1.0.2 in /usr/local/lib/python3.7/dist-packages (from spacy->bert-extractive-summarizer) (1.0.5)\n",
            "Requirement already satisfied: preshed<3.1.0,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from spacy->bert-extractive-summarizer) (3.0.5)\n",
            "Requirement already satisfied: requests<3.0.0,>=2.13.0 in /usr/local/lib/python3.7/dist-packages (from spacy->bert-extractive-summarizer) (2.23.0)\n",
            "Requirement already satisfied: thinc==7.4.0 in /usr/local/lib/python3.7/dist-packages (from spacy->bert-extractive-summarizer) (7.4.0)\n",
            "Requirement already satisfied: murmurhash<1.1.0,>=0.28.0 in /usr/local/lib/python3.7/dist-packages (from spacy->bert-extractive-summarizer) (1.0.5)\n",
            "Requirement already satisfied: blis<0.5.0,>=0.4.0 in /usr/local/lib/python3.7/dist-packages (from spacy->bert-extractive-summarizer) (0.4.1)\n",
            "Requirement already satisfied: plac<1.2.0,>=0.9.6 in /usr/local/lib/python3.7/dist-packages (from spacy->bert-extractive-summarizer) (1.1.3)\n",
            "Requirement already satisfied: catalogue<1.1.0,>=0.0.7 in /usr/local/lib/python3.7/dist-packages (from spacy->bert-extractive-summarizer) (1.0.0)\n",
            "Requirement already satisfied: cymem<2.1.0,>=2.0.2 in /usr/local/lib/python3.7/dist-packages (from spacy->bert-extractive-summarizer) (2.0.5)\n",
            "Requirement already satisfied: importlib-metadata; python_version < \"3.8\" in /usr/local/lib/python3.7/dist-packages (from transformers->bert-extractive-summarizer) (3.8.1)\n",
            "Requirement already satisfied: regex!=2019.12.17 in /usr/local/lib/python3.7/dist-packages (from transformers->bert-extractive-summarizer) (2019.12.20)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.7/dist-packages (from transformers->bert-extractive-summarizer) (20.9)\n",
            "Requirement already satisfied: tokenizers<0.11,>=0.10.1 in /usr/local/lib/python3.7/dist-packages (from transformers->bert-extractive-summarizer) (0.10.1)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.7/dist-packages (from transformers->bert-extractive-summarizer) (3.0.12)\n",
            "Requirement already satisfied: sacremoses in /usr/local/lib/python3.7/dist-packages (from transformers->bert-extractive-summarizer) (0.0.43)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.13.0->spacy->bert-extractive-summarizer) (2020.12.5)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.13.0->spacy->bert-extractive-summarizer) (2.10)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.13.0->spacy->bert-extractive-summarizer) (1.24.3)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.13.0->spacy->bert-extractive-summarizer) (3.0.4)\n",
            "Requirement already satisfied: typing-extensions>=3.6.4; python_version < \"3.8\" in /usr/local/lib/python3.7/dist-packages (from importlib-metadata; python_version < \"3.8\"->transformers->bert-extractive-summarizer) (3.7.4.3)\n",
            "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.7/dist-packages (from importlib-metadata; python_version < \"3.8\"->transformers->bert-extractive-summarizer) (3.4.1)\n",
            "Requirement already satisfied: pyparsing>=2.0.2 in /usr/local/lib/python3.7/dist-packages (from packaging->transformers->bert-extractive-summarizer) (2.4.7)\n",
            "Requirement already satisfied: click in /usr/local/lib/python3.7/dist-packages (from sacremoses->transformers->bert-extractive-summarizer) (7.1.2)\n",
            "Requirement already satisfied: six in /usr/local/lib/python3.7/dist-packages (from sacremoses->transformers->bert-extractive-summarizer) (1.15.0)\n",
            "Installing collected packages: bert-extractive-summarizer\n",
            "Successfully installed bert-extractive-summarizer-0.7.1\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "sRN0DjXr2jwH",
        "outputId": "f6b09bb7-e537-4d32-96db-3f5c3a31cdd9"
      },
      "source": [
        "pip install sumy"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting sumy\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/61/20/8abf92617ec80a2ebaec8dc1646a790fc9656a4a4377ddb9f0cc90bc9326/sumy-0.8.1-py2.py3-none-any.whl (83kB)\n",
            "\r\u001b[K     |████                            | 10kB 4.5MB/s eta 0:00:01\r\u001b[K     |███████▉                        | 20kB 7.7MB/s eta 0:00:01\r\u001b[K     |███████████▊                    | 30kB 9.5MB/s eta 0:00:01\r\u001b[K     |███████████████▋                | 40kB 7.9MB/s eta 0:00:01\r\u001b[K     |███████████████████▌            | 51kB 4.3MB/s eta 0:00:01\r\u001b[K     |███████████████████████▍        | 61kB 5.1MB/s eta 0:00:01\r\u001b[K     |███████████████████████████▍    | 71kB 4.7MB/s eta 0:00:01\r\u001b[K     |███████████████████████████████▎| 81kB 5.2MB/s eta 0:00:01\r\u001b[K     |████████████████████████████████| 92kB 3.4MB/s \n",
            "\u001b[?25hCollecting breadability>=0.1.20\n",
            "  Downloading https://files.pythonhosted.org/packages/ad/2d/bb6c9b381e6b6a432aa2ffa8f4afdb2204f1ff97cfcc0766a5b7683fec43/breadability-0.1.20.tar.gz\n",
            "Requirement already satisfied: requests>=2.7.0 in /usr/local/lib/python3.7/dist-packages (from sumy) (2.23.0)\n",
            "Collecting pycountry>=18.2.23\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/76/73/6f1a412f14f68c273feea29a6ea9b9f1e268177d32e0e69ad6790d306312/pycountry-20.7.3.tar.gz (10.1MB)\n",
            "\u001b[K     |████████████████████████████████| 10.1MB 14.6MB/s \n",
            "\u001b[?25hRequirement already satisfied: docopt<0.7,>=0.6.1 in /usr/local/lib/python3.7/dist-packages (from sumy) (0.6.2)\n",
            "Requirement already satisfied: nltk>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from sumy) (3.2.5)\n",
            "Requirement already satisfied: chardet in /usr/local/lib/python3.7/dist-packages (from breadability>=0.1.20->sumy) (3.0.4)\n",
            "Requirement already satisfied: lxml>=2.0 in /usr/local/lib/python3.7/dist-packages (from breadability>=0.1.20->sumy) (4.2.6)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests>=2.7.0->sumy) (1.24.3)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests>=2.7.0->sumy) (2.10)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests>=2.7.0->sumy) (2020.12.5)\n",
            "Requirement already satisfied: six in /usr/local/lib/python3.7/dist-packages (from nltk>=3.0.2->sumy) (1.15.0)\n",
            "Building wheels for collected packages: breadability, pycountry\n",
            "  Building wheel for breadability (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for breadability: filename=breadability-0.1.20-py2.py3-none-any.whl size=21680 sha256=6ee0a09815115cedcfee02f49c7ac9a909ead7fff932ecab92186e9fd18dafd7\n",
            "  Stored in directory: /root/.cache/pip/wheels/5a/4d/a1/510b12c5e65e0b2b3ce539b2af66da0fc57571e528924f4a52\n",
            "  Building wheel for pycountry (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for pycountry: filename=pycountry-20.7.3-py2.py3-none-any.whl size=10746863 sha256=726ee6ce6eaaeaf98223780315709e300df37c572c09dcda1892304621e45f3b\n",
            "  Stored in directory: /root/.cache/pip/wheels/33/4e/a6/be297e6b83567e537bed9df4a93f8590ec01c1acfbcd405348\n",
            "Successfully built breadability pycountry\n",
            "Installing collected packages: breadability, pycountry, sumy\n",
            "Successfully installed breadability-0.1.20 pycountry-20.7.3 sumy-0.8.1\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "FP75Gbtm7o-l",
        "outputId": "64c72523-4da1-46e0-e16e-c28bdbda1ea1"
      },
      "source": [
        "import nltk\n",
        "nltk.download('punkt')"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[nltk_data] Downloading package punkt to /root/nltk_data...\n",
            "[nltk_data]   Unzipping tokenizers/punkt.zip.\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "True"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "XCmOh3JX97UI",
        "outputId": "90ea4722-2089-4ec0-c579-5f5973a67404"
      },
      "source": [
        "pip install requests"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: requests in /usr/local/lib/python3.7/dist-packages (2.23.0)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests) (1.24.3)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests) (3.0.4)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests) (2.10)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests) (2020.12.5)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Ve6PuCVu9_Ab",
        "outputId": "725369d0-683d-42f4-c12e-9ebcc4e20d5a"
      },
      "source": [
        "pip install bs4"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: bs4 in /usr/local/lib/python3.7/dist-packages (0.0.1)\n",
            "Requirement already satisfied: beautifulsoup4 in /usr/local/lib/python3.7/dist-packages (from bs4) (4.6.3)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Gy7QulmUgv5x",
        "outputId": "83782b29-fc7e-481e-caca-c4cb1a46171b"
      },
      "source": [
        "pip install PyPDF2"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting PyPDF2\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/b4/01/68fcc0d43daf4c6bdbc6b33cc3f77bda531c86b174cac56ef0ffdb96faab/PyPDF2-1.26.0.tar.gz (77kB)\n",
            "\r\u001b[K     |████▎                           | 10kB 16.3MB/s eta 0:00:01\r\u001b[K     |████████▌                       | 20kB 19.1MB/s eta 0:00:01\r\u001b[K     |████████████▊                   | 30kB 9.3MB/s eta 0:00:01\r\u001b[K     |█████████████████               | 40kB 7.9MB/s eta 0:00:01\r\u001b[K     |█████████████████████▏          | 51kB 4.4MB/s eta 0:00:01\r\u001b[K     |█████████████████████████▍      | 61kB 5.0MB/s eta 0:00:01\r\u001b[K     |█████████████████████████████▋  | 71kB 5.3MB/s eta 0:00:01\r\u001b[K     |████████████████████████████████| 81kB 3.5MB/s \n",
            "\u001b[?25hBuilding wheels for collected packages: PyPDF2\n",
            "  Building wheel for PyPDF2 (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for PyPDF2: filename=PyPDF2-1.26.0-cp37-none-any.whl size=61085 sha256=497d2eef15c6dc286e4eaa34e0f9a42a14b19f8da4198155b6e27f122e0401d6\n",
            "  Stored in directory: /root/.cache/pip/wheels/53/84/19/35bc977c8bf5f0c23a8a011aa958acd4da4bbd7a229315c1b7\n",
            "Successfully built PyPDF2\n",
            "Installing collected packages: PyPDF2\n",
            "Successfully installed PyPDF2-1.26.0\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "gMmcZCB414Sn",
        "outputId": "01c9671e-2b3b-4104-a280-34660e1b5786"
      },
      "source": [
        "pip install newspaper3k"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting newspaper3k\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/d7/b9/51afecb35bb61b188a4b44868001de348a0e8134b4dfa00ffc191567c4b9/newspaper3k-0.2.8-py3-none-any.whl (211kB)\n",
            "\u001b[K     |████████████████████████████████| 215kB 4.5MB/s \n",
            "\u001b[?25hRequirement already satisfied: Pillow>=3.3.0 in /usr/local/lib/python3.7/dist-packages (from newspaper3k) (7.1.2)\n",
            "Collecting feedparser>=5.2.1\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/1c/21/faf1bac028662cc8adb2b5ef7a6f3999a765baa2835331df365289b0ca56/feedparser-6.0.2-py3-none-any.whl (80kB)\n",
            "\u001b[K     |████████████████████████████████| 81kB 3.9MB/s \n",
            "\u001b[?25hRequirement already satisfied: beautifulsoup4>=4.4.1 in /usr/local/lib/python3.7/dist-packages (from newspaper3k) (4.6.3)\n",
            "Requirement already satisfied: python-dateutil>=2.5.3 in /usr/local/lib/python3.7/dist-packages (from newspaper3k) (2.8.1)\n",
            "Requirement already satisfied: PyYAML>=3.11 in /usr/local/lib/python3.7/dist-packages (from newspaper3k) (3.13)\n",
            "Requirement already satisfied: lxml>=3.6.0 in /usr/local/lib/python3.7/dist-packages (from newspaper3k) (4.2.6)\n",
            "Requirement already satisfied: nltk>=3.2.1 in /usr/local/lib/python3.7/dist-packages (from newspaper3k) (3.2.5)\n",
            "Collecting jieba3k>=0.35.1\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/a9/cb/2c8332bcdc14d33b0bedd18ae0a4981a069c3513e445120da3c3f23a8aaa/jieba3k-0.35.1.zip (7.4MB)\n",
            "\u001b[K     |████████████████████████████████| 7.4MB 7.0MB/s \n",
            "\u001b[?25hCollecting tinysegmenter==0.3\n",
            "  Downloading https://files.pythonhosted.org/packages/17/82/86982e4b6d16e4febc79c2a1d68ee3b707e8a020c5d2bc4af8052d0f136a/tinysegmenter-0.3.tar.gz\n",
            "Requirement already satisfied: requests>=2.10.0 in /usr/local/lib/python3.7/dist-packages (from newspaper3k) (2.23.0)\n",
            "Collecting tldextract>=2.0.1\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/7e/62/b6acd3129c5615b9860e670df07fd55b76175b63e6b7f68282c7cad38e9e/tldextract-3.1.0-py2.py3-none-any.whl (87kB)\n",
            "\u001b[K     |████████████████████████████████| 92kB 9.1MB/s \n",
            "\u001b[?25hCollecting cssselect>=0.9.2\n",
            "  Downloading https://files.pythonhosted.org/packages/3b/d4/3b5c17f00cce85b9a1e6f91096e1cc8e8ede2e1be8e96b87ce1ed09e92c5/cssselect-1.1.0-py2.py3-none-any.whl\n",
            "Collecting feedfinder2>=0.0.4\n",
            "  Downloading https://files.pythonhosted.org/packages/35/82/1251fefec3bb4b03fd966c7e7f7a41c9fc2bb00d823a34c13f847fd61406/feedfinder2-0.0.4.tar.gz\n",
            "Collecting sgmllib3k\n",
            "  Downloading https://files.pythonhosted.org/packages/9e/bd/3704a8c3e0942d711c1299ebf7b9091930adae6675d7c8f476a7ce48653c/sgmllib3k-1.0.0.tar.gz\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.7/dist-packages (from python-dateutil>=2.5.3->newspaper3k) (1.15.0)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests>=2.10.0->newspaper3k) (2.10)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests>=2.10.0->newspaper3k) (2020.12.5)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests>=2.10.0->newspaper3k) (3.0.4)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests>=2.10.0->newspaper3k) (1.24.3)\n",
            "Collecting requests-file>=1.4\n",
            "  Downloading https://files.pythonhosted.org/packages/77/86/cdb5e8eaed90796aa83a6d9f75cfbd37af553c47a291cd47bc410ef9bdb2/requests_file-1.5.1-py2.py3-none-any.whl\n",
            "Requirement already satisfied: filelock>=3.0.8 in /usr/local/lib/python3.7/dist-packages (from tldextract>=2.0.1->newspaper3k) (3.0.12)\n",
            "Building wheels for collected packages: jieba3k, tinysegmenter, feedfinder2, sgmllib3k\n",
            "  Building wheel for jieba3k (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for jieba3k: filename=jieba3k-0.35.1-cp37-none-any.whl size=7398406 sha256=8f2e30d522bc1b12944781407941f33e86474a1346ef1e3551995ef4f2790a63\n",
            "  Stored in directory: /root/.cache/pip/wheels/83/15/9c/a3f1f67e7f7181170ad37d32e503c35da20627c013f438ed34\n",
            "  Building wheel for tinysegmenter (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for tinysegmenter: filename=tinysegmenter-0.3-cp37-none-any.whl size=13538 sha256=91025fe74b8d82d3c9fec021a420eaee23b0022497c4996ac738af75441f53d7\n",
            "  Stored in directory: /root/.cache/pip/wheels/81/2b/43/a02ede72324dd40cdd7ca53aad718c7710628e91b8b0dc0f02\n",
            "  Building wheel for feedfinder2 (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for feedfinder2: filename=feedfinder2-0.0.4-cp37-none-any.whl size=3358 sha256=ac192d94e65d87ed735eaaf2b7d1e921fd86863dc2d3d1a8f4986b5c508c3c17\n",
            "  Stored in directory: /root/.cache/pip/wheels/de/03/ca/778e3a7a627e3d98836cc890e7cb40c7575424cfd3340f40ed\n",
            "  Building wheel for sgmllib3k (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for sgmllib3k: filename=sgmllib3k-1.0.0-cp37-none-any.whl size=6067 sha256=d6cdad31c99cd4b48f87115ea81d9480f1aaa437ee765213f1d0c5db022ee1b2\n",
            "  Stored in directory: /root/.cache/pip/wheels/f1/80/5a/444ba08a550cdd241bd9baf8bae44be750efe370adb944506a\n",
            "Successfully built jieba3k tinysegmenter feedfinder2 sgmllib3k\n",
            "Installing collected packages: sgmllib3k, feedparser, jieba3k, tinysegmenter, requests-file, tldextract, cssselect, feedfinder2, newspaper3k\n",
            "Successfully installed cssselect-1.1.0 feedfinder2-0.0.4 feedparser-6.0.2 jieba3k-0.35.1 newspaper3k-0.2.8 requests-file-1.5.1 sgmllib3k-1.0.0 tinysegmenter-0.3 tldextract-3.1.0\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "OPNzZFIe3inX"
      },
      "source": [
        "# Libraries"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "9V-V8T-B3hG9"
      },
      "source": [
        "from nltk.cluster.util import cosine_distance\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import networkx as nx\n",
        "from nltk.tokenize import sent_tokenize\n",
        "from sentence_transformers import SentenceTransformer\n",
        "import requests\n",
        "from bs4 import BeautifulSoup"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Tvs6BI6_CrYr"
      },
      "source": [
        "from sumy.parsers.plaintext import PlaintextParser\n",
        "from sumy.nlp.tokenizers import Tokenizer"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "jmkwmkRU3u53"
      },
      "source": [
        "from sklearn.metrics.pairwise import cosine_similarity\n",
        "from sklearn.metrics.pairwise import euclidean_distances\n",
        "from sumy.parsers.plaintext import PlaintextParser\n",
        "from sumy.nlp.tokenizers import Tokenizer\n",
        "import PyPDF2\n",
        "from PyPDF2 import PdfFileReader\n",
        "from pdfminer.high_level import extract_text\n",
        "import newspaper "
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "m2zokUYX5xT_"
      },
      "source": [
        "import requests\n",
        "#import RAKE\n",
        "import operator\n",
        "import timeit\n",
        "import time\n",
        "API_KEY = \"AIzaSyAvZlAZHcE3lXigI2ZJxWQvCoCpF6qFiyY\"\n",
        "SEARCH_ENGINE_ID = \"bf6585a79f0c6d8bb\""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7J0vWULO5ZVj"
      },
      "source": [
        "# Functions"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "TecHuT-o5mAN"
      },
      "source": [
        "def google_search(api_key, search_engine, query, page_number=1):\n",
        "  start = (page_number - 1) * 10 + 1\n",
        "  url = f\"https://www.googleapis.com/customsearch/v1?key={api_key}&cx={search_engine}&q={query}&start={start}\"\n",
        "  data = requests.get(url).json()\n",
        "  return data"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "RXHtpvgz5mFl"
      },
      "source": [
        "def results(data):\n",
        "  search_items = data.get(\"items\")\n",
        "  for i, search_item in enumerate(search_items, start=1):\n",
        "    title = search_item.get(\"title\")\n",
        "    snippet = search_item.get(\"snippet\")\n",
        "    link = search_item.get(\"link\")\n",
        "    print(\"Title:\", title)\n",
        "    print(\"Description:\", snippet)\n",
        "    print(\"URL:\", link, \"\\n\")\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ZVdNMBX06C2R"
      },
      "source": [
        "def only_link(data):\n",
        "  search_items = data.get(\"items\")\n",
        "  link=[]\n",
        "  for i, search_item in enumerate(search_items, start=1):\n",
        "    link.append(search_item.get(\"link\")) \n",
        "    #print(\"URL:\", link, \"\\n\")\n",
        "  return  link"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "B9mAN3WF6Kxf"
      },
      "source": [
        "def Sort (info):\n",
        "  info.sort(key = lambda x: x[1], reverse=True)\n",
        "  return info"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ld6WkGV36QTB"
      },
      "source": [
        "def read_file(files):\n",
        "  myfile = open(files, \"r\")\n",
        "  readfiles=myfile.readlines()\n",
        "  text =\"\"\n",
        "  for line in range(len(readfiles)):\n",
        "    text = text + \" \"+ readfiles[line]\n",
        "  return text "
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "uheL9FSh6jXL"
      },
      "source": [
        "# Actual Using LSA Summarizer"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "0AMfae9H64Tn"
      },
      "source": [
        "textatten = read_file(\"moralsandethics.txt\")\n",
        "sentencesAtten=sent_tokenize(textatten)"
      ],
      "execution_count": 53,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0zHQX69F7-7h",
        "outputId": "8cc69072-6eb2-4fe3-e613-d89e280103f4"
      },
      "source": [
        "from sumy.summarizers.lsa import LsaSummarizer\n",
        "parser = PlaintextParser.from_string(textatten,Tokenizer(\"english\"))\n",
        "summarizer_lsa = LsaSummarizer()\n",
        "# Summarize using sumy LSA\n",
        "summary =summarizer_lsa(parser.document,2)\n",
        "lsa_summary=\"\"\n",
        "for sentence in summary:\n",
        "    lsa_summary+=str(sentence)  \n",
        "print(lsa_summary)"
      ],
      "execution_count": 54,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "There are some matters the law will be agnostic on but which ethics and morality have a lot to say.Secondly, there may be times when obeying the law would require us to act against our ethics or morality.\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "SwQLncTFX98S",
        "outputId": "3a4491a2-96dc-448b-f9f3-33dfb62ea51b"
      },
      "source": [
        "type(lsa_summary)"
      ],
      "execution_count": 33,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "str"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 33
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Ztvj_l_3YdSV"
      },
      "source": [
        "stringgg=\"Accumulated mutations were very strongly enriched for functional categories, which combined with the high levels of convergence observed suggest that mutation accumulation was governed by strong positive selection (Avrani, et al. 2017).While overall mutation accumulation rates within mutators are substantially higher, the mutations accumulated tend to be enriched for transitions, which have a lower likelihood of disrupting protein sequences.\""
      ],
      "execution_count": 36,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rhKV2-j78vN-",
        "outputId": "ec22700c-1e85-4a23-d34d-31c62b503e32"
      },
      "source": [
        "#myquery= input(\"Please enter the information to be searched \")\n",
        "data =google_search(API_KEY,SEARCH_ENGINE_ID,lsa_summary)\n",
        "results(data)\n",
        "#print(data)"
      ],
      "execution_count": 55,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Title: Ethics, morality, law – what's the difference? - The Ethics Centre\n",
            "Description: Sep 27, 2016 ... Some people talk about their personal ethics, others talk about a set of morals, \n",
            "and everyone in a ... There are some matters the law will be agnostic on but \n",
            "which ethics and morality have a lot to say. ... Secondly, there may be times when \n",
            "obeying the law would require us to act against our ethics or morality.\n",
            "URL: https://ethics.org.au/ethics-morality-law-whats-the-difference/ \n",
            "\n",
            "Title: What is Ethics? - Markkula Center for Applied Ethics\n",
            "Description: Jan 1, 2010 ... \"Ethics has to do with what my feelings tell me is right or wrong.\" \"Ethics ... Like \n",
            "Baumhart's first respondent, many people tend to equate ethics with their feelings\n",
            ". But being ethical is clearly not a matter of following one's feelings. A person ... \n",
            "But laws, like feelings, can deviate from what is ethical. Our own ...\n",
            "URL: https://www.scu.edu/ethics/ethics-resources/ethical-decision-making/what-is-ethics/ \n",
            "\n",
            "Title: Divine Command Theory | Internet Encyclopedia of Philosophy\n",
            "Description: On a law conception of ethics, conformity with the virtues requires obeying the ... \n",
            "God's commands do not determine morality, but rather inform us about its content\n",
            ". ... Moreover, Kant argues that “there is not the slightest ground in the moral law \n",
            "for a ... Some might object that this is overly egoistic, but at any rate it seems less\n",
            " ...\n",
            "URL: https://iep.utm.edu/divine-c/ \n",
            "\n",
            "Title: MORALITY AND THE WILL OF GOD James Fieser 9/1/2017 ...\n",
            "Description: The second option is that good things are objectively good, and God merely ... \n",
            "For God can do anything that is not self-contradictory or act in any way that does \n",
            "... Scotus concludes that God could alter virtually all of the moral laws if he \n",
            "wanted. ... But as the believer heaps more and more powerful abilities on God, \n",
            "there is a ...\n",
            "URL: https://www.utm.edu/staff/jfieser/class/300/god.htm \n",
            "\n",
            "Title: The Natural Law Tradition in Ethics (Stanford Encyclopedia of ...\n",
            "Description: Sep 23, 2002 ... It might be thought that there is nothing that can be done to begin a discussion of \n",
            "... If any moral theory is a theory of natural law, it is Aquinas's. ... all of them we \n",
            "can say that they are in the neighborhood of the natural law view but ... The \n",
            "second is that, when we focus on the human's role as recipient of the ...\n",
            "URL: https://plato.stanford.edu/entries/natural-law-ethics/ \n",
            "\n",
            "Title: Euthyphro dilemma - Wikipedia\n",
            "Description: The Euthyphro dilemma is found in Plato's dialogue Euthyphro, in which Socrates \n",
            "asks ... Thus Euthyphro's theory does not give us the very nature of the pious, but \n",
            "at most a quality ... God is bound by the laws of morality instead of being their \n",
            "establisher. ... First, there are some divine commands that can directly create \n",
            "moral ...\n",
            "URL: https://en.wikipedia.org/wiki/Euthyphro_dilemma \n",
            "\n",
            "Title: Moral Reasoning (Stanford Encyclopedia of Philosophy)\n",
            "Description: Sep 15, 2003 ... While moral reasoning can be undertaken on another's behalf, it is ... we also \n",
            "reason theoretically about what morality requires of us; but the nature ... On the \n",
            "one side, there is the first-order question of what moral truths there are, if any. ... \n",
            "and action – has a lot to say to such questions, both in its traditional, ...\n",
            "URL: https://plato.stanford.edu/entries/reasoning-moral/ \n",
            "\n",
            "Title: Mere Christianity By C.S. Lewis\n",
            "Description: to believe that there is one God and that Jesus Christ is His only Son. Finally ... I \n",
            "was not writing to expound something I could call \"my religion,\" but to expound \"\n",
            "mere\" ... To say more would take me at once into highly controversial regions. ... \n",
            "But at those moments when we are most conscious of the Moral Law, it usually.\n",
            "URL: https://www.dacc.edu/assets/pdfs/PCM/merechristianitylewis.pdf \n",
            "\n",
            "Title: How Oxford and Peter Singer drove me from atheism to Jesus - The ...\n",
            "Description: May 22, 2017 ... I didn't need faith to ground my identity or my values. ... Awesome is His Love to \n",
            "any soul that can experience Him. ... But if there is a god, and he designed the \n",
            "system, I would say he has ... Secondly, while we act from a very limited moral \n",
            "and temporal ... God has intervened many times in human history.\n",
            "URL: http://www.veritas.org/oxford-atheism-to-jesus/ \n",
            "\n",
            "Title: The Human Basis Of Laws And Ethics - American Humanist ...\n",
            "Description: Mar 24, 2021 ... There is a tendency on the part of many theists to assume that the burden of proof \n",
            "is ... the second assumption — that the source of moral values must lie outside of \n",
            "... But what is so often forgotten is that those human laws that appear externally ... \n",
            "(1) If we can obey our own traffic laws without the need of a ...\n",
            "URL: https://americanhumanist.org/what-is-humanism/human-basis-laws-ethics/ \n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ebG1Xh358jy3"
      },
      "source": [
        "# WebStripping "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tQ19QhXXZaa8",
        "outputId": "e10e7ccd-9e44-4974-99eb-0f6260a6d97c"
      },
      "source": [
        "URL_link =only_link(data)\n",
        "Extracted_pdf =[]\n",
        "length=0\n",
        "if len(URL_link)>2:\n",
        "  length =2\n",
        "else:\n",
        "  length = len(URL_link)\n",
        "for i in range(length):\n",
        "  if URL_link[i].find(\"pdf\")!=-1:\n",
        "    response = requests.get(URL_link[i]) \n",
        "    pdf = open(\"pdf\"+str(i)+\".pdf\", 'wb') \n",
        "    pdf.write(response.content) \n",
        "    pdf.close() \n",
        "    print(\"File \", i, \" downloaded\")\n",
        "    text = extract_text(\"pdf\"+str(i)+\".pdf\")\n",
        " \n",
        "    #print(text)\n",
        "    Extracted_pdf.append(text)\n",
        "  \n",
        "  else:\n",
        "    Url_i = newspaper.Article(url=\"%s\" % (URL_link[i]), language='en') \n",
        "    Url_i.download() \n",
        "    Url_i.parse() \n",
        "    Extracted_pdf.append(Url_i.text)\n",
        "    print('Successful')"
      ],
      "execution_count": 56,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Successful\n",
            "Successful\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "uOiLy7YFv9kp"
      },
      "source": [
        "sentencesweb=sent_tokenize(text)\n",
        "sentencesAtten=sent_tokenize(textatten)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Im99WuYuplLX"
      },
      "source": [
        "# Document Similarity"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "maDRNNfJr2Sc"
      },
      "source": [
        "Tokenized_Stripped=[]\n",
        "for i in range(len(Extracted_pdf)):\n",
        "  sentenceh=sent_tokenize(Extracted_pdf[i])\n",
        "  Tokenized_Stripped.append(sentenceh)\n"
      ],
      "execution_count": 57,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 312,
          "referenced_widgets": [
            "0595cc45a97d4436a419113d34d82b36",
            "e8a5f6a41f4244368a1e5cad5bca7b25",
            "8703c475a615485f8358aec5634854c6",
            "9defc53806e24e57b8ee9188efb2f438",
            "1dd85dfd78d64ce2bcd4339d250d7d43",
            "6bb25d52fb19411ba1ff5710bc165a9a",
            "9b4397d6c26f4d62808c6772b29a2f11",
            "af0a28394d034f4882051a1c945b0272",
            "87e75f2764fa41908bc106cc5c2a68bf",
            "a7e610734d6544bfa6f82b0e0f272110",
            "39b061cb06bd4fe6a2eb675f2987e253",
            "817cb5cc10894bb9943b48d89861c9a5",
            "47d57a057d734cea8f01f5f857bc903e",
            "c9c8573e5f7c4b63bf8d20707580b81e",
            "99288fe0d15341128db52f71d9536ee3",
            "b2b0a8e1cc6d46e18b499009af7230a1",
            "4864f597528a455690b3786b61b81a78",
            "fa8b785a948447178b5695857a9852a1",
            "c67964d3c9c549f9ba3e5fae610e47f0",
            "cd75a3ae6682496085415b48ed3a7038",
            "e60cae3ceed54230b6d53c5af2505351",
            "06065f123a474ef3aae7529522d3ac01",
            "a3df8c95d38e413cad50b64ff253252e",
            "87323962ea2e4866a82de3e9a3ec0888",
            "0ac253bc93ac4d27b1f29dfe1eeb5332",
            "45c5d6b64352490db3b4266395d4fa44",
            "53b970bc82c44adca9289066e4991b56",
            "37bf76bec6a34d0a92a7887eef085a3b",
            "d4f9540213ad40e5836ffb0e5aeefbc6",
            "f60fa13ade1e4141a6859871af99539d",
            "135de97544d34137bcb2278876347a73",
            "da3e6e15544a489c9acc0c80a8f12b8c"
          ]
        },
        "id": "hOtOzEEmdx2w",
        "outputId": "619b2c8f-01ab-4a3c-dd3b-33ed7ad53a5b"
      },
      "source": [
        "model = SentenceTransformer('bert-base-nli-mean-tokens')\n",
        "for i in range(len(Tokenized_Stripped)):\n",
        "  text_embeddings = model.encode(Tokenized_Stripped[i], batch_size = 8, show_progress_bar = True)\n",
        "  text_embeddings2 = model.encode(sentencesAtten, batch_size = 8, show_progress_bar = True)\n",
        "  np.shape(text_embeddings)\n",
        "  np.shape(text_embeddings2)\n",
        "  similarities = cosine_similarity(text_embeddings,text_embeddings2)\n",
        "  similarities_sorted = similarities.argsort()\n",
        "  id_1 = []\n",
        "  id_2 = []\n",
        "  score = []\n",
        "  for index,array in enumerate(similarities_sorted):\n",
        "    id_1.append(index)\n",
        "    id_2.append(array[-2])\n",
        "    score.append(similarities[index][array[-2]])\n",
        "    index_df = pd.DataFrame({'id_1' : id_1,\n",
        "                          'id_2' : id_2,\n",
        "                          'score' : score})\n",
        "  sum_of_sims =(np.sum(index_df[\"score\"]))\n",
        "  percentage_of_similarity = round(float((sum_of_sims / len(Tokenized_Stripped[i])) * 100))\n",
        "  print(f'Average similarity float: {float(sum_of_sims / len(Tokenized_Stripped[i]))}')\n",
        "  print(f'Average similarity percentage: {float(sum_of_sims / len(Tokenized_Stripped[i])) * 100}')\n",
        "  print(f'Average similarity rounded percentage: {percentage_of_similarity}')\n"
      ],
      "execution_count": 58,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.jupyter.widget-view+json": {
              "model_id": "0595cc45a97d4436a419113d34d82b36",
              "version_minor": 0,
              "version_major": 2
            },
            "text/plain": [
              "HBox(children=(FloatProgress(value=0.0, description='Batches', max=4.0, style=ProgressStyle(description_width=…"
            ]
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "stream",
          "text": [
            "\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.jupyter.widget-view+json": {
              "model_id": "87e75f2764fa41908bc106cc5c2a68bf",
              "version_minor": 0,
              "version_major": 2
            },
            "text/plain": [
              "HBox(children=(FloatProgress(value=0.0, description='Batches', max=4.0, style=ProgressStyle(description_width=…"
            ]
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "stream",
          "text": [
            "\n",
            "Average similarity float: 0.7565378392896345\n",
            "Average similarity percentage: 75.65378392896345\n",
            "Average similarity rounded percentage: 76\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.jupyter.widget-view+json": {
              "model_id": "4864f597528a455690b3786b61b81a78",
              "version_minor": 0,
              "version_major": 2
            },
            "text/plain": [
              "HBox(children=(FloatProgress(value=0.0, description='Batches', max=6.0, style=ProgressStyle(description_width=…"
            ]
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "stream",
          "text": [
            "\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.jupyter.widget-view+json": {
              "model_id": "0ac253bc93ac4d27b1f29dfe1eeb5332",
              "version_minor": 0,
              "version_major": 2
            },
            "text/plain": [
              "HBox(children=(FloatProgress(value=0.0, description='Batches', max=4.0, style=ProgressStyle(description_width=…"
            ]
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "stream",
          "text": [
            "\n",
            "Average similarity float: 0.6842834613424666\n",
            "Average similarity percentage: 68.42834613424667\n",
            "Average similarity rounded percentage: 68\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "NbHdCv65atmr",
        "outputId": "8fc6ae32-5f13-48c2-a3e7-9f811ae256d8"
      },
      "source": [
        "similarities = cosine_similarity(text_embeddings,text_embeddings2)\n",
        "similarities_sorted = similarities.argsort()\n",
        "id_1 = []\n",
        "id_2 = []\n",
        "score = []\n",
        "for index,array in enumerate(similarities_sorted):\n",
        "    id_1.append(index)\n",
        "    id_2.append(array[-2])\n",
        "    score.append(similarities[index][array[-2]])\n",
        "index_df = pd.DataFrame({'id_1' : id_1,\n",
        "                          'id_2' : id_2,\n",
        "                          'score' : score})\n",
        "sum_of_sims =(np.sum(index_df[\"score\"]))\n",
        "percentage_of_similarity = round(float((sum_of_sims / len(sentencesweb)) * 100))\n",
        "print(f'Average similarity float: {float(sum_of_sims / len(sentencesweb))}')\n",
        "print(f'Average similarity percentage: {float(sum_of_sims / len(sentencesweb)) * 100}')\n",
        "print(f'Average similarity rounded percentage: {percentage_of_similarity}')"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "239.74783873558044\n",
            "Average similarity float: 0.8154688392366681\n",
            "Average similarity percentage: 81.54688392366681\n",
            "Average similarity rounded percentage: 82\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "0lZO4PgCqR4W"
      },
      "source": [
        "sentencesweb=sent_tokenize(text)\n",
        "sentencesAtten=sent_tokenize(textatten)\n",
        "model = SentenceTransformer('bert-base-nli-mean-tokens')\n",
        "text_embeddings = model.encode(sentencesweb, batch_size = 8, show_progress_bar = True)\n",
        "text_embeddings2 = model.encode(sentencesAtten, batch_size = 8, show_progress_bar = True)\n",
        "np.shape(text_embeddings)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "JT3UdPcex1GQ",
        "outputId": "a034cc75-0f21-4cf9-87dc-0f2a1d5c5105"
      },
      "source": [
        "percentage_of_similarity = round(float((sum_of_sims / len(sentencesweb)) * 100))\n",
        "print(f'Average similarity float: {float(sum_of_sims / len(sentencesweb))}')\n",
        "print(f'Average similarity percentage: {float(sum_of_sims / len(sentencesweb)) * 100}')\n",
        "print(f'Average similarity rounded percentage: {percentage_of_similarity}')"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Average similarity float: 0.8154688392366681\n",
            "Average similarity percentage: 81.54688392366681\n",
            "Average similarity rounded percentage: 82\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XwHVB_9e9HSd"
      },
      "source": [
        "# Trial 2"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "CwVeVAuk88Vm"
      },
      "source": [
        "#Tokenizing the received\n",
        "textevol = read_file(\"Evolution.txt\")\n",
        "sentencesevol=sent_tokenize(textevol)\n",
        "textStruct = read_file(\"Structural_bio.txt\")\n",
        "sentencesStruct=sent_tokenize(textStruct)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "bzKAAHQi_7DA",
        "outputId": "6a62e563-d199-46d1-ca7f-354eb69bba88"
      },
      "source": [
        "parser = PlaintextParser.from_string(textStruct,Tokenizer(\"english\"))\n",
        "summarizer_lsa = LsaSummarizer()\n",
        "# Summarize using sumy LSA\n",
        "summary =summarizer_lsa(parser.document,2)\n",
        "lsa_summary=\"\"\n",
        "for sentence in summary:\n",
        "    lsa_summary+=str(sentencesStruct)  \n",
        "print(lsa_summary)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[' Structural bioinformatics\\n CLoNe: automated clustering based on local density\\n neighborhoods for application to biomolecular structural\\n ensembles\\n Sylvain Tra¨ ger1,2, Giorgio Tamo` 1,2, Deniz Aydin1,2, Giulia Fonti1,2,\\n Martina Audagnotto1,2 and Matteo Dal Peraro1,2,*\\n 1\\n Institute of Bioengineering, School of Life Sciences, Ecole Polytechnique Fe´ de´ rale de Lausanne, Lausanne 1025, Switzerland and 2\\n Swiss Institute of Bioinformatics, Lausanne 1015, Switzerland\\n *To whom correspondence should be addressed.', 'Associate Editor: Lenore Cowen\\n Received on January 8, 2020; revised on July 14, 2020; editorial decision on August 12, 2020; accepted on August 18, 2020\\n Abstract\\n Motivation: Proteins are intrinsically dynamic entities.', 'Flexibility sampling methods, such as molecular dynamics or\\n those arising from integrative modeling strategies, are now commonplace and enable the study of molecular conformational landscapes in many contexts.', 'Resulting structural ensembles increase in size as technological and algorithmic advancements take place, making their analysis increasingly demanding.', 'In this regard, cluster analysis\\n remains a go-to approach for their classification.', 'However, many state-of-the-art algorithms are restricted to specific\\n cluster properties.', 'Combined with tedious parameter fine-tuning, cluster analysis of protein structural ensembles\\n suffers from the lack of a generally applicable and easy to use clustering scheme.', 'Results: We present CLoNe, an original Python-based clustering scheme that builds on the Density Peaks algorithm\\n of Rodriguez and Laio.', 'CLoNe relies on a probabilistic analysis of local density distributions derived from nearest\\n neighbors to find relevant clusters regardless of cluster shape, size, distribution and amount.', 'We show its capabilities on many toy datasets with properties otherwise dividing state-of-the-art approaches and improves on the original algorithm in key aspects.', 'Applied to structural ensembles, CLoNe was able to extract meaningful conformations from membrane binding events and ligand-binding pocket opening as well as identify dominant dimerization\\n motifs or inter-domain organization.', 'CLoNe additionally saves clusters as individual trajectories for further analysis\\n and provides scripts for automated use with molecular visualization software.', 'Availability and implementation: www.epfl.ch/labs/lbm/resources, github.com/LBM-EPFL/CLoNe.', 'Contact: matteo.dalperaro@epfl.ch\\n Supplementary information: Supplementary data are available at Bioinformatics online.', '1 Introduction\\n The perception of molecular structures, especially proteins, is gradually shifting from the concept of one single and rigid structure to\\n the idea that biomolecules natively exhibit a continuum of states\\n (Frank, 2018).', 'Protein folding, post-translational modifications\\n (Audagnotto and Dal Peraro, 2017), binding to other molecules or\\n their involvement in catalytic events result in vast and complex conformational landscapes.', 'Molecular dynamics (MD), thanks to progress in both its technological and algorithmic aspects, allows for\\n the simulation of key biomolecular events.', 'Their observability, however, tends to be limited by currently accessible timescales.', 'Researchers consistently come up with innovative protocols to push\\n this limit further (Barducci et al., 2011; Bussi, 2014; Chavent et al.,\\n 2016; Doerr et al., 2016; Hamelberg et al., 2004; Noe´ et al., 2019;\\n Shirts and Pande, 2000; Sultan et al., 2018; Wassenaar et al., 2015)\\n granting us with the ability to capture protein folding as well as protein–protein, protein–membrane and protein–ligand interactions\\n (Audagnotto et al., 2016; De Vivo et al., 2016; McKiernan et al.,\\n 2017; Oleinikovas et al., 2016).', 'State-of-the-art protocols for smallmolecule docking (Amaro et al., 2018; Kokh et al., 2011; Vahl\\n Quevedo et al., 2014), protein–protein docking and integrative\\n modeling strategies, in general, have shifted toward the integration\\n of dynamics in some form as well (Abriata and Dal Peraro, 2020;\\n Malhotra et al., 2019; Tamo` et al., 2015).', 'All of the aforementioned\\n aspects advocate dynamics as a cornerstone of modern structural\\n biology and push the need for efficient tools to extract functional insight from structural ensembles in general.', 'VC The Author(s) 2020.', 'Published by Oxford University Press.', '1\\n This is an Open Access article distributed under the terms of the Creative Commons Attribution Non-Commercial License (http://creativecommons.org/licenses/by-nc/4.0/),\\n which permits non-commercial re-use, distribution, and reproduction in any medium, provided the original work is properly cited.', 'For commercial re-use, please contact\\n journals.permissions@oup.com\\n Bioinformatics, 2020, 1–8\\n doi: 10.1093/bioinformatics/btaa742\\n Advance Access Publication Date: 21 August 2020\\n Original Paper\\n Downloaded from https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa742/5895303 by guest on 19 March 2021\\n However, these advances come at a price.', 'The sheer size, the intrinsic complexity and redundancy of structural ensembles makes\\n their successful analysis and computational integration non-trivial.', 'Coarse-graining tools such as cluster analysis effectively reduce simulations of thousands of conformations to few key biological states\\n and hence constitute a go-to approach with countless applications to\\n date [(Cheng et al., 2008; De Paris et al., 2015; de Souza et al.,\\n 2017), reviewed in (Peng et al., 2018; Shao et al., 2007)].', 'Such states\\n may serve as basis of Markov state models (Husic and Pande, 2018;\\n Wang et al., 2018).', 'To our knowledge, however, an algorithm able\\n to cluster data efficiently irrespective of their properties is still missing.', 'Indeed, different cluster shapes, sizes and densities usually dictate which clustering approach is best suited for a given task.', 'The\\n most known and widely used scheme is probably that of k-means\\n and its many variations (Jain, 2010).', 'This center-based clustering\\n scheme, however, suffers from unequal results due to random initializations, the a priori setting of the number of clusters and its limitation to spherical clusters.', 'Alternatively, hierarchical schemes such as\\n the Ward-linkage agglomerative algorithm (Ward, 1963) do not require pre-setting the number of clusters and are popular for building\\n Markov state models (Beauchamp et al., 2012; Husic and Pande,\\n 2017; Paris et al., 2015).', 'They are however sensitive to noise and\\n outliers and may suffer from non-spherical clusters (Peng et al.,\\n 2018).', 'Conversely, DBSCAN (Ester et al., 1996) is able to manage\\n clusters regardless of shape by utilizing density differences between\\n clusters and noise.', 'However, setting its parameter is not trivial and\\n its optimal value may not be unique throughout the dataset when\\n clusters of largely different densities are present.', 'This limitation is at\\n the core of OPTICS, which can be seen as an extension of DBSCAN\\n (Ankerst et al., 1999), although not strictly advertised as a clustering\\n algorithm.', 'Defining metastable states of proteins is non-trivial due to the\\n large and often redundant number of internal degrees of freedom,\\n yielding sampled conformational spaces with local minima often devoid of biological significance.', 'We can make the assumption that,\\n given enough sampling and a choice of relevant features, metastable\\n states would lie in regions or clusters of high density, which would\\n be separated by valleys of different density levels that would correspond to transitional states.', 'Furthermore, no assumption can be\\n made on the shape or relative densities of clusters, which would depend on both conformational sampling and target system.', 'Rodriguez and Laio (2014) designed the Density Peaks (DP) algorithm aimed at clustering regardless of shape and dimensionality.', 'Their algorithm generated significant interest thanks to their clever\\n definition of cluster centers, which states that a cluster center should\\n display a higher density (q) than its neighbors and a high distance to\\n another point of higher density (d).', 'DP takes a single input parameter, which relates to a cutoff distance for the computation of q.', 'However, it requires the user to specify thresholds for both q and d\\n mid-computation in order to select the cluster centers, which prevents a fully automated clustering process.', 'DP has since been\\n improved by the inclusion of k nearest neighbors (kNN) (Du et al.,\\n 2016; Xie et al., 2016; Zhang and Li, 2015) or heat diffusion\\n (Mehmood et al., 2016) for a more robust estimation of q, which\\n allows for a better handling of cases where clusters have significantly different densities.', 'These improvements still require user intervention for selecting cluster centers.', 'Conversely, Wang and Xu (2017)\\n built on DP to automatically select cluster centers based on maximizing an average silhouette index, although other input parameters\\n are required instead.', 'Liang and Cheng coupled principles from\\n DBSCAN with a divide-and-conquer approach to recursively and\\n automatically select cluster centers (Liang and Chen, 2016).', 'Recently, d’Errico et al.', 'coupled DP with a non-parametric density\\n estimator (Rodriguez et al., 2018), yielding Density Peaks Advanced\\n (DPA; d’Errico et al., 2018).', 'While exhibiting impressive robustness\\n to a variety of cluster shapes, densities and to outliers, DPA still suffers from a few issues.', 'We found that it performed worse than the\\n original on some typical benchmark datasets, and requires a sensitive albeit unique input parameter.', 'Moreover, both DP and DPA exhibit an inconsistent outlier removal procedure.', 'These drawbacks\\n may prove crucial when targeting structural biological data, where\\n regions at lower effective density may have equal or even increased\\n significance than others at higher densities.', 'The complexity of biological structures leads to numerous unique yet equally relevant\\n choices of features, each with their own topology.', 'The analysis of\\n such datasets is greatly hindered by sensitivity to input parameters,\\n which implies that tedious fine-tuning steps have to be undertaken.', 'Here, we introduce an approach to remedy these drawbacks,\\n enabling a facilitated analysis of complex real-world datasets from\\n structural biology.', 'Our approach builds on the original DP algorithm by introducing a fragmenting of the data into specific density\\n distributions.', 'In essence, the local densities of each point are computed using nearest neighbors and a Gaussian kernel and points\\n associated with local density maxima are identified as putative cluster centers.', 'To increase robustness to non-spherical cluster shapes,\\n clusters are merged using the Bhattacaryaa coefficient\\n (Bhattacharyya, 1943) by comparing density distributions derived\\n from putative cluster cores and boundaries.', 'Finally, outliers from\\n impromptu noise fluctuations are removed by means of a Bayes classifier.', 'This, to the best of our knowledge, constitute an original contribution to the density peaks algorithm.', 'Termed Clustering based\\n on Local density Neighborhoods (CLoNe), our approach relies on a\\n single input parameter that is both robust and intuitive to set.', 'We\\n test it on many typical benchmark datasets and against state-of-theart clustering schemes.', 'The local focus of CLoNe allows for the detection of biological states of smaller frequency while its ease of use\\n allows the researcher to focus on choosing relevant biological features for pre-processing or analyzing their structural ensemble without being hindered by algorithmic limitations.', 'Furthermore, CLoNe\\n outputs useful molecular visualization scripts for the validation of\\n cluster relevance in the target biological context (Supplementary Fig.', 'S1).', 'We apply CLoNe on a range of structural datasets from MD\\n simulations or integrative modeling studies, each time detailing the\\n feature selection process and which information can be extracted\\n from the results.', 'Our examples cover previously published studies\\n on protein–membrane interactions, internal structural rearrangements of disordered proteins, cryptic allosteric pocket formation and\\n transmembrane dimerization motifs, and highlight the broad advantages of CLoNe for the analysis of molecular structural ensembles.', '2 Materials and methods\\n An overview and basic usage of CLoNe is available in\\n Supplementary Figure S1, on GitHub and the webpage of our laboratory (see Abstract section).', 'We created a synthetic dataset containing clusters of significantly different densities and various shapes\\n in order to showcase the procedure behind our approach.', 'CLoNe\\n starts by finding the k nearest neighbors of each point in a dataset X\\n of N points using k-nearest neighbors (kNN) (Pedregosa et al.,\\n 2011), yielding a neighbor matrix M where each row i contains all\\n the neighbors j of point i in increasing order of Euclidean distance.', 'To account for significant density differences between clusters, we\\n initially assume that all points are cluster centers.', 'In a first step, we\\n estimate the local density q of each point i using a Gaussian kernel:\\n qi ¼ X\\n jkNNi\\n e\\n  Mi;j\\n dc\\n  2\\n ¼ X\\n jkNNi\\n qij (1)\\n where kNNi is the set of nearest neighbors of i in increasing order of\\n distance and dc is a cutoff distance defined as to be superior to a\\n user-defined percentage pdc (the single input parameter of CLoNe)\\n of all distances within M, similar to the original DP algorithm\\n (Rodriguez and Laio, 2014).', 'We define the core of putative cluster i\\n as the set of neighbors that contribute to qi at least as much of the\\n j1 previous neighbors on average:\\n corei ¼ j  kNNijqij\\n 1\\n pdcN\\n Xj1\\n k¼0 qik \\x04 \\x05 (2)\\n We show in Figure 1a the cardinality (number of elements) of\\n the core of each point in our synthetic dataset.', 'As expected, this\\n 2 S.Tra¨ ger et al.', 'Downloaded from https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa742/5895303 by guest on 19 March 2021\\n number is higher for points closer to real cluster centers and lower\\n for points laying on the outskirt of clusters.', 'The visualization of core\\n cardinalities is an efficient way to observe the underlying topology\\n of the dataset.', 'In order to identify if i is a genuine candidate for cluster center, we identify its first neighbor j of higher density.', 'If neighbor j belongs to corei, then neighbor j is a better candidate for\\n cluster center in this region than i. Conversely, i is a genuine candidate for cluster center if j is not in the core of i.', 'Cluster assignation\\n is done in a single step by assigning a point to the same cluster as its\\n nearest point of higher density, in order of decreasing local density,\\n similar to the original DP approach.', 'The results at this stage of\\n CLoNe are shown in Figure 1a.', 'One of the drawbacks of the DP algorithm is its limited ability to\\n deal with clusters with more than one peak or those with an elongated region of similar density, such as the noisy circles benchmark\\n dataset (Supplementary Fig.', 'S2).', 'This is true at this stage of CLoNe\\n as well, as the Gaussian kernel in (1) is biased toward determining\\n cores with spherical shapes (Fig.', '1b).', 'We can make the assumptions\\n that if two existing clusters A and B should be merged, then the\\n density from one core to the other should be relatively constant.', 'This can be estimated by looking at the core cardinality distribution\\n of the points belonging to the core of both clusters as well as that of\\n the points from the boundary between them (Fig.', '1b, left), which\\n can be defined as:\\n boundaryAB ¼ i 2 A; j 2 Bjd ið Þ ; j < dcg \\x06 (3)\\n Then, we define the following probability density function for\\n the ensemble of points belonging to either cluster cores or the\\n boundary from (3):\\n Ps ¼ KDE corei ð Þ f g ; i 2 S (4)\\n where S denotes one of the aforementioned ensembles and #corei the\\n core cardinality of point i and KDE refers to the probability density\\n function estimated by unimodal Gaussian kernel density estimation.', 'Similarity between probability distributions can be measured using\\n the Bhattacharyya coefficient (BC) (Bhattacharyya, 1943), which is\\n bound between 0 and 1.', 'Thus, the formula to compute the BC between the core of cluster A with the boundary from (3) becomes:\\n BCA ¼ ffiffiffiffiffiffiffiffiffiffiffiffiffiffiffiffiffiffiffiffiffiffiffiffi\\n PboundPcoreA\\n p (5)\\n We take the decision of merging Clusters A and B if the mean of\\n their respective BC with their boundary is above a threshold sBC:\\n B Að Þ ; B ¼ 1\\n 2 ð Þ BCA þ BCB > sBC (6)\\n where PcoreB is obtained as for cluster A in (5) and sBC was chosen to\\n be the 65th percentile to limit uncertainty and based on benchmarks\\n (Fig.', '1 and Supplementary Fig.', 'S2).', 'Taking the mean of both coefficients prevents the merging of a cluster whose probability density is\\n similar to that of its boundary with a cluster of significantly higher\\n density.', 'This enabled us to identify clusters that can hardly be\\n defined with a single density peak, such as uniform density over\\n non-spherical shapes (Figs 1b and 2a).', 'The point with highest q is\\n chosen as the new cluster center.', 'If pdc is chosen too small, clusters\\n may be split into sub-clusters.', 'Within our approach, these subclusters are likely to be merged into clusters matching the original\\n topology, expanding the range of acceptable values for pdc and making it an input parameter less sensitive than the one of DPA\\n (Supplementary Fig.', 'S3).', 'The second drawback of using the Gaussian kernel in (1) is that\\n it may falsely identify impromptu local noise fluctuations as cluster\\n centers (Fig.', '1b).', 'To remedy this, we define two probability density\\n functions.', 'Mcores is the probability density of the local density of all\\n points belonging to any cluster core as per (2):\\n Mcores ¼ KDEð qx f g ; x 2 corei 8i 2 C Þ (7)\\n where C is the set of cluster centers remaining after the previous\\n merging step.', 'Moutliers is the equivalent function for all points identified as outliers:\\n Moutliers ¼ KDEð qx; x 2 X if qx < f \\x04 qcenterx\\n \\x06 \\x08Þ (8)\\n where qcenterx is the local density of the center of the cluster x\\n belongs to and f an arbitrary fraction chosen to be 0.1 by default\\n and used throughout this article.', 'To determine if an identified cluster is more probable to be derived from noise than to be a genuine\\n cluster, we use a Bayesian classifier.', 'For each cluster core ci, we derive the following posterior probabilities using Bayes’ theorem:\\n p Yð Þ¼ jci\\n p Yð Þ pðcijYÞ\\n pðciÞ ; Y 2 f g outliers; cores (9)\\n where the prior probabilities are defined as follows:\\n Fig.', '1.', 'Clustering based on local density neighborhoods.', 'We created a synthetic\\n dataset that contains four croissant-shaped clusters of 500 elements each with different scaling as well as a cluster in the shape of a cross with 4000 elements and added\\n Gaussian noise.', '(a) On the left, the cardinality of the core of each point in the dataset.', 'On the right, the clusters obtained after the first stage of clustering.', '(b) The average Bhattacaryya coefficient B(A,B) between two Clusters A and B is shown.', 'The\\n upper left plots show an example of two clusters being merged corresponding to the\\n upper square in panel (a).', 'The lower left plots show an example of two clusters that\\n will not be merged corresponding to the lower square in panel (a).', 'Points belonging\\n to the core of each cluster are shown in black, regular points in shades of gray and\\n points belonging to the boundary in red.', 'The plot on the right shows the clustering\\n after merging clusters.', '(c) CLoNe uses a Bayesian classifier to decide if a cluster is\\n genuine or arises from noise fluctuations.', 'On the left, the corresponding probability\\n density functions of points belonging to any cluster cores (black), noise (blue), or to\\n individual cluster cores (light blue).', 'The range of local density of clusters classified\\n as noise fluctuations are shown on the secondary y-axis (dark blue).', 'The final clustering result is shown on the right, with identified outliers shown as black crosses\\n CLoNe: automated clustering of biomolecular structural ensembles 3\\n Downloaded from https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa742/5895303 by guest on 19 March 2021\\n p Yð Þ ¼ jYj\\n N ; Y 2 f g outliers; cores (10)\\n where jYj denotes the cardinality of the corresponding class.', 'The\\n likelihoods pðcijYÞ can be computed by evaluating the previously\\n defined probability distributions (7) and (8) at ci.', 'Disregarding the\\n evidence pðciÞ common for both outliers and cores classes, we thus\\n obtain the following Bayesian classifier:\\n y^i ¼ argmax Y2f g outliers; cores\\n p Yð ÞY\\n x2ci\\n MYðqxÞ (11)\\n This classifier enabled us to remove all cluster centers arising\\n from noise fluctuations.', 'Combined with the previous merging step,\\n the clustering is now complete (Fig.', '1c).', 'The software has been written using Python3.7.', 'When used on\\n structural biological data, CLoNe outputs cluster centers as separate\\n PDB files, individual clusters as XTC trajectories and Tcl scripts for\\n automated loading within the visualization software VMD (Visual\\n Molecular Dynamics, (Humphrey et al., 1996)).', 'The loading of MD\\n trajectories as well as the saving of cluster centers and cluster subtrajectories is done through the MDTraj package (McGibbon et al.,\\n 2015).', 'We use the scipy (Jones et al., 2001), scikit-learn (Pedregosa\\n et al., 2011) and Statsmodels (Seabold and Perktold, 2010) packages\\n for many operations described in the previous subsection and to\\n compare CLoNe to other clustering algorithms.', '3 Results\\n 3.1 Automatic cluster center determination from local\\n density neighborhood analysis\\n We applied CLoNe to a large set of common benchmark datasets\\n from various sources (Chang and Yeung, 2008; d’Errico et al., 2018;\\n Fra¨nti and Sieranoja, 2018; Fu and Medico, 2007; Gionis et al.,\\n 2007; Pedregosa et al., 2011), covering different key properties of\\n clusters, such as non-spherical shapes, anisotropy, as well as significant size and density differences, all of which can be expected from\\n real-world datasets from structural biology.', 'In the previous section,\\n we detailed how CLoNe automatically detects cluster centers, accurately merges clusters and removes outliers, succeeding in cases where\\n previous iterations of DP did not (Fig.', '2a and Supplementary Fig.', 'S2).', 'Similarly to the original DP algorithm (Rodriguez and Laio,\\n 2014), CLoNe requires a single input parameter pdc, which relates\\n to a cutoff distance used in the estimation of local densities (see\\n Section 2).', 'In general, pdc takes a value in a small range and is\\n intuitive to set.', 'For instance, with pdc values of 1 or 2 local densities\\n will be estimated considering neighborhoods small enough to identify individual spiral branches as clusters (Fig.', '2b).', 'For higher values\\n of pdc, the scale of the Gaussian kernel in Equation (1) will increase\\n and merge individual branches into whole spirals, allowing the study\\n of multiple hierarchies intuitively (Fig.', '2c).', 'Other than clusters with\\n non-spherical shapes, CLoNe identifies successfully the numerous\\n Gaussian clusters of the A3 dataset (Fig.', '2d).', 'Some degree of overlapping in real-world datasets is to be expected.', 'The S4 dataset contains 15 highly overlapping Gaussian clusters of varying densities\\n and shapes but equal size.', 'As with the A3 dataset, CLoNe does not\\n perform unnecessary merging with nearby clusters (Fig.', '2e) and is\\n robust to large amounts of outliers on top of clusters with significantly different densities (Fig.', '2f).', 'This general applicability of\\n CLoNe coupled with a single, robust and easy to set input parameter\\n (Supplementary Fig.', 'S3) is unique among the commonly used clustering algorithms found in the Scikit-learn package (Pedregosa et al.,\\n 2011).', 'In fact, CLoNe is among the fastest algorithms from that\\n package in addition to both DP and DPA and the most accurate on\\n the available benchmark cases (Supplementary Fig.', 'S2).', 'The only\\n other algorithm succeeding on all benchmark cases is OPTICS\\n (Ankerst et al., 1999), which runs slightly slower than CLoNe on\\n these datasets and tends to classify too many points as outliers.', 'Similar to the original implementation of DP, CLoNe is applicable\\n to high dimensionality datasets as well (Supplementary Table S1\\n and Supplementary Fig.', 'S4).', 'One of the principal aims of this work is to offer a clustering algorithm able to classify unlabeled biological structural ensembles\\n into relevant states associated with their function and mechanism of\\n action.', 'We have applied CLoNe to real-world structural biology\\n data reporting on the dynamic conformational space of a protein\\n that associates with its specific biological membrane, cryptic allosteric pocket opening and dimerization of transmembrane proteins.', '3.2 Determining relevant states within protein\\n conformational ensembles\\n COQ9 is a lipid-binding protein associated with the biosynthesis of\\n coenzyme Q (CoQ), a redox-active lipid that is essential for cellular\\n respiration (Lohman et al., 2014).', 'Recently, coarse-grained molecular dynamics (CG-MD) simulations and liposome co-flotation assays\\n were used together to reveal that COQ9 accesses membranes in a\\n multi-step fashion through a distinct, C-terminal amphipathic helix\\n (a10) (Lohman et al., 2019).', 'In these simulations, COQ9 first diffused in the aqueous environment, then underwent various\\n Fig.', '2.', 'Local neighborhood density analysis for automated center and cluster determination.', 'For each panel, the cardinality of the core of each point as detailed in the Section 2\\n is shown on the left.', 'The resulting clusters are shown on the right, with clusters in different colors and identified outliers as black crosses.', '(a) Noisy circles dataset.', '(b and c)\\n Four instances of spiral dataset with different values of the input parameter yielding different yet valid clusters.', '(d) The a3 dataset containing 50 Gaussian clusters.', '(e) The s4\\n dataset with highly overlapping Gaussian clusters, some with anisotropic distributions.', '(f) A synthetic dataset with clusters of significantly different sizes and densities taken\\n from Density Peaks Advanced\\n 4 S.Tra¨ ger et al.', 'Downloaded from https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa742/5895303 by guest on 19 March 2021\\n conformational changes upon membrane binding (Lohman et al.,\\n 2019).', 'We applied CLoNe to the CG-MD trajectory used in the latter study and sought to identify the main binding events pertaining\\n to the protein itself.', 'To this end, we extracted features characterizing both its movements in the aqueous environment through monitoring its distance to the membrane as well as key conformational\\n changes based on the angle between the unique a10 helix of COQ9\\n and its globular domain (Fig.', '3a, left).', 'Using these two features,\\n CLoNe outputs three clusters, each of which seem to follow\\n Gaussian distributions (Fig.', '3b).', 'One cluster regroups all conformations that correspond to diffusion movements in the aqueous environment, while the other two highlight the membrane association of\\n a10 first followed by the globular domain as a converging step\\n (Fig.', '3a), thus its higher density (Fig.', '3b).', 'Similar results can be\\n obtained hypothesis-free by using raw atomic spatial coordinates\\n (Supplementary Fig.', 'S4).', 'The Gaussian distribution of structural\\n clusters has also been observed in a recent study from our group\\n involving the KAP1 protein, where CLoNe was also successfully\\n applied (Supplementary Fig.', 'S5) (Fonti et al., 2019).', '3.2 Isolating sub-ensembles of relevant conformations\\n for ligand–target interactions\\n In recent years, small-molecule docking software is no stranger to\\n dynamics, taking into account ensembles of ligand conformations\\n (Amaro et al., 2018) or receptor flexibility (Kokh et al., 2011;\\n Salmaso and Moro, 2018; Vahl Quevedo et al., 2014).', 'A recent\\n study highlighted a novel replica exchange-based MD protocol combined with benzene probes, where each replica harbors a different\\n scaling of water–protein interactions (Oleinikovas et al., 2016).', 'Using this method, the authors could observe the opening of cryptic\\n allosteric pockets in several systems, including that of the TEM1 blactamase, which plays a critical role in antibiotic resistance (Horn\\n and Shoichet, 2004).', 'The simulations were started from the apo\\n crystal structure with a closed allosteric pocket (Fig.', '4a).', 'Out of the\\n eight replicas of the simulation, we chose three with neutral (first),\\n medium (fourth) and highest (last) scaling factors as a tradeoff between maximizing the sampled conformational space and limiting\\n redundancy of the over-represented closed conformations\\n (Oleinikovas et al., 2016) (Fig.', '4b).', 'Along with key residue R244 on\\n the opposite wall of the pocket, the opening of a-helices H11 and\\n H12 and key residues L220 and N276 dictate pocket opening and\\n allow two inhibitors to be accommodated (Horn and Shoichet,\\n 2004), while the three mentioned residues form a triad when the\\n pocket is closed.', 'In addition to the opening of the two helices, visual\\n inspection of the simulations indicated a deepening of the pocket.', 'As\\n a result, we chose features tracking the distance between the Ca of\\n residues L220 and N276 as well as that of their sidechains to monitor pocket opening as well as the distance between the Ca of I263\\n and I279 as a measure of pocket depth (Fig.', '4a).', 'The original study\\n used fpocket (Le Guilloux et al., 2009) to monitor pocket exposure\\n in each replica (Fig.', '4b, top).', 'The same was done on the clusters\\n obtained by CLoNe (Fig.', '4b, bottom), showing different levels of\\n pocket openness.', 'Corresponding cluster centers highlight key structural differences between each state (Fig.', '4c), which are representative of the feature distribution per cluster (Fig.', '4d, top).', 'Cluster\\n assignation follows the observation of the original publication,\\n where open states were more prevalent in the replica of medium\\n scaling (Fig.', '4d, bottom).', '3.3 Identifying dominant conformational motifs in\\n protein oligomerization\\n Another challenge in structural biology is the understanding of how\\n biomolecules oligomerize to distinctive functional states.', 'One of\\n these cases, the transmembrane a-helix of the Amyloid Precursor\\n Protein (termed APP hereafter), has recently been studied by our lab\\n through the high-throughput MD protocol DAFT (Docking Assay\\n For Transmembrane components, (Wassenaar et al., 2015)) in order\\n to identify which of two dimerization motifs is promoted depending\\n on the lipid composition of the synaptic plasma membrane\\n (Audagnotto et al., 2016).', 'The G700G704G708 motif is thought to\\n direct the binding of APP to regulators promoting cholesterol biosynthesis, while the G709A713 motif would bind to cholesterol molecules (Fig.', '5a).', 'Extracting features from molecular datasets is not\\n always straightforward.', 'Macromolecular movements possess an inherent redundancy due to the sheer number of internal degrees of\\n freedom or prior knowledge may be lacking in order to select meaningful features, such as those highlighted in Figures 3 and 4.', 'The use\\n of dimensionality reduction methods, such as principal component\\n analysis (PCA) has been seen for clustering of MD simulations\\n (Wolf and Kirschner, 2013) and can help identifying coordinates of\\n significance while discarding less useful dimensions.', 'The DAFT simulations of APP from (Audagnotto et al., 2016) are over 2 ms in total\\n and contain countless states, many corresponding to unbound\\n monomers.', 'The first principal component based on the Cartesian\\n coordinates of the coarse grain backbone covers 77% of the variability in the simulation, highlighting two clusters (Fig.', '5b).', 'The\\n blue cluster of lower amplitude corresponds to all states exhibiting\\n unbound monomers (Fig.', '5e, left), while the second cluster regroups\\n all the dimerized states regardless of motif.', 'Focusing on that cluster,\\n we calculated the pair-wise distances between the backbone atoms\\n of each motifs in both helices (Fig.', '5a) and reduced these features to\\n a two-dimensional principal space covering 94% of the variability\\n before clustering (Fig.', '5c).', 'We want to highlight CLoNe’s ability to\\n analyze the neighborhood of low-density clusters without influence\\n from high-density regions (Fig.', '5c and Supplementary Table S2).', 'Clusters in blue in Figure 5d all depict states close to the\\n G700G704G708 motifs and those in green the G709A713 motif.', 'In the\\n middle are two clusters, shown in brown, that we interpret as hybrid.', 'In all cases, the darker-shaded clusters of each group correspond to the closest to the optimal motif arrangement, while the\\n others can be considered as closely related metastable states\\n (Supplementary Fig.', 'S6).', 'Similar to the original study (Audagnotto\\n Fig.', '3.', 'Utilizing Gaussian cluster properties to extract centers as key biological states of the COQ9 membrane protein.', '(a) COQ9 and its associated features, which include an\\n internal angle h and its distance to the membrane d (left).', 'Cluster centers are shown on the right side of the panel.', '(b) Every frame is plotted in the mentioned feature space and\\n color coded according to their core cardinality (left) and cluster assignation (right), which follows the same color code as in (a).', 'Outliers are shown as black crosses and centers\\n as black stars\\n CLoNe: automated clustering of biomolecular structural ensembles 5\\n Downloaded from https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa742/5895303 by guest on 19 March 2021\\n et al., 2016), CLoNe finds the preferred dimerization motif to be\\n G700G704G708 as evidenced by the corresponding centers’ local densities, cluster population and core cardinality (Fig.', '5c and\\n Supplementary Table S2).', 'A comparison between CLoNe and other\\n state-of-the-art algorithms on all the structural data shown in this\\n section demonstrates the advantages of CLoNe for analysis of molecular structural ensembles (Supplementary Fig.', 'S7).', '4 Discussion\\n Many clustering methods rely on parameters that are often nontrivial to optimize or on random initial conditions that may drastically change the outcome.', 'Commonly used algorithms are generally\\n restricted to specific cluster properties, forcing the researcher\\n through a process of trial and error.', 'Moreover, choosing relevant\\n features from structural datasets is challenging and different features\\n may generate different cluster topologies, sometimes irrelevant.', 'CLoNe was designed with these issues in mind and aims to provide\\n a stream-lined analytic process to yield results rapidly along with\\n helpful visualization scripts to analyze and confirm the relevance of\\n the clusters in the target biological context (Supplementary Fig.', 'S1).', 'CLoNe’s only parameter regulates the size of the local neighborhood\\n considered around each data point, which can be regarded as cluster\\n sizing parameter.', 'Its value need only be decreased if clusters seem\\n too inclusive and vice-versa, making CLoNe an intuitive algorithm\\n to use in addition to its general applicability.', 'For structural datasets,\\n CLoNe is able to extract clusters as separate trajectories and provides scripts for their automatic loading in the visualization software\\n VMD (Humphrey et al., 1996).', 'For larger macromolecules, the concept of a conformational state is blurry, hard to determine and often\\n depends on context.', 'It is not always clear which features to use to\\n obtain an accurate partition of the structural ensemble.', 'The results\\n obtained on COQ9 can be obtained hypothesis-free on raw spatial\\n coordinates or using PCA to extract relevant features\\n (Supplementary Fig.', 'S4).', 'This was done in the case of APP as well as\\n to reduce an otherwise redundant feature space to one of lower\\n dimensionality.', 'If one wishes to disentangle internal from overall\\n motion, dihedral PCA was used with success to study peptide folding (Altis et al., 2007; Mu et al., 2004).', 'However, when other\\n features than the chosen ones can be expected to exhibit motions of\\n larger amplitudes, PCA will favor the latter over the former.', 'This is\\n true for the TEM1 b-lactamase, where internal structural motions\\n will be more prevalent than the fluctuations of the selected key\\n pocket residues.', 'In such cases, a feature-based approach is to be preferred.', 'Alternatively, some will advocate the use of time-lagged independent component analysis (TICA) (Naritomi and Fuchigami,\\n 2011) instead.', 'TICA was found to be the better alternative for building Markov state models (Husic and Pande, 2018; Pe´rez-Herna´ndez\\n et al., 2013).', 'However, in cases where large amplitude fluctuations\\n are the target or when there is redundancy in features, we believe\\n that PCA remains a safe approach.', 'As the conformational ensembles presented in this study tend to\\n exhibit Gaussian distributions, CLoNe may thus be used to extract\\n cluster centers as higher probability states.', 'Such states offer an overview of the ensemble and may serve as starting models for building\\n Markov state models in general.', 'Moreover, the precision of the classification achieved by CLoNe enables the identification of dominant\\n biological states from large datasets.', 'Beyond the case of APP,\\n CLoNe identified different key pocket conformations in the case of\\n TEM1 b-lactamase.', 'Further clustering efforts on this system should\\n target the different positions of R244, which was not tracked in this\\n study but was previously shown to play a dual role between TEM1’s\\n active site and allosteric pocket (Horn and Shoichet, 2004).', 'CLoNe\\n may then be used as a pre-processing tool prior to small-molecule\\n docking studies, where accounting for receptor flexibility is an active\\n development area (Kokh et al., 2011; Vahl Quevedo et al., 2014).', 'Integrative modeling aims at incorporating data from multiple\\n sources to determine the structure of macromolecular complexes.', 'Such hybrid strategies typically combine low resolution data of\\n whole complexes with high resolution structures of their components so as to predict the quaternary structure of the former\\n (Cassidy et al., 2018).', 'This process is however severely hindered by\\n structural dynamics differing between a complex and its isolated\\n components.', 'For this reason, many hybrid modeling strategies now\\n incorporate some form of dynamics to bridge this gap (Malhotra\\n et al., 2019; Tamo` et al., 2015).', 'While we previously utilized classical MD for the prediction of heptameric aerolysin pores (Degiacomi\\n et al., 2013; Degiacomi and Dal Peraro, 2013), such an approach\\n would not be feasible for heteromultimeric assemblies where\\n Fig.', '4.', 'Identification of different opening states of the allosteric cryptic pocket in TEM1 b-lactamase.', '(a) apo and Holo structures (left and right, respectively).', 'Allosteric inhibitors are shown in gray and white.', 'Features following helical opening include the distance between Ca atoms of N276 and L220 (medium blue) and the Cc of their sidechain\\n (light blue).', 'Pocket depth is monitored by the distance between Ca-carbons of I263 and 279 (dark blue).', '(b) The pocket exposure calculated using the fpocket software for the\\n original replicas (top) and for each clusters (bottom).', 'The dotted line in both is the reference value of the holo crystal structure used in the original paper.', '(c) The center of each\\n cluster in cartoon representation on top of a surface representation of the allosteric pocket, highlighting the different states of helical openness and pocket depth.', 'The triad\\n N276-L220-R244 governing pocket opening and closing are shown as gray sticks.', '(c) The distribution of each feature for each cluster (top) and the cluster assignation along\\n the three chosen replicas (bottom)\\n 6 S.Tra¨ ger et al.', 'Downloaded from https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa742/5895303 by guest on 19 March 2021\\n multiple conformational ensembles are required simultaneously.', 'Reducing them to their crucial components may enable the structural characterization of large macromolecular complexes, which\\n may otherwise be intractable.', 'Applied to the fields of smallmolecule docking, integrative modeling and structural dynamics\\n studies, CLoNe presents itself as a versatile and powerful tool for\\n modern computational structural biology.', 'Acknowledgements\\n We thank Vladimiras Oleinikovas and Francesco L. Gervasio for providing\\n the simulations of the TEM1 b-lactamase as well as offering general advice;\\n Lucien F. Krapp and Romain Groux for helpful discussions.', 'Funding\\n M.D.P.', 'lab was supported by the Swiss National Science Foundation (grants\\n number 200021_157217 and 31003A_170154).', 'Conflict of Interest: none declared.', 'References\\n Abriata,L.A.', 'and Dal Peraro,M.', '(2020) Will cryo-electron microscopy shift\\n the current paradigm in protein structure prediction?', 'J. Chem.', 'Inf.', 'Model.,\\n 60, 2443–2447.', 'Altis,A.', 'et al.', '(2007) Dihedral angle principal component analysis of molecular\\n dynamics simulations.', 'J. Chem.', 'Phys., 126, 244111.', 'Amaro,R.E.', 'et al.', '(2018) Ensemble docking in drug discovery.', 'Biophys.', 'J.,\\n 114, 2271–2278.', 'Ankerst,M.', 'et al.', '(1999) OPTICS: Ordering Points to Identify the Clustering\\n Structure.', 'In: Proceedings of the 1999 ACM SIGMOD International\\n Conference on Management of Data, SIGMOD ’99, pp.', '49–60.', 'ACM, New\\n York, NY, USA.', 'Audagnotto,M.', 'and Dal Peraro,M.', '(2017) Protein post-translational modifications: in silico prediction tools and molecular modeling.', 'Comput.', 'Struct.', 'Biotechnol.', 'J., 15, 307–319.', 'Audagnotto,M.', 'et al.', '(2016) Effect of the synaptic plasma membrane on the\\n stability of the amyloid precursor protein homodimer.', 'J. Phys.', 'Chem.', 'Lett.,\\n 7, 3572–3578.', 'Barducci,A.', 'et al.', '(2011) Metadynamics.', 'Wiley Interdiscip.', 'Rev.', 'Comput.', 'Mol.', 'Sci., 1, 826–843.', 'Beauchamp,K.A.', 'et al.', '(2012) Simple few-state models reveal hidden complexity in protein folding.', 'Proc.', 'Natl.', 'Acad.', 'Sci.', 'USA, 109, 17807–17813.', 'Bhattacharyya,A.', '(1943) On a measure of divergence between two statistical\\n populations defined by their probability distributions.', 'Bull.', 'Calcutta Math.', 'Soc., 35, 99–109.', 'Bussi,G.', '(2014) Hamiltonian replica exchange in GROMACS: a flexible implementation.', 'Mol.', 'Phys., 112, 379–384.', 'Cassidy,C.K.', 'et al.', '(2018) CryoEM-based hybrid modeling approaches for\\n structure determination.', 'Curr.', 'Opin.', 'Microbiol., 43, 14–23.', 'Chang,H.', 'and Yeung,D.-Y.', '(2008) Robust path-based spectral clustering.', 'Pattern Recogn., 41, 191–203.', 'Chavent,M.', 'et al.', '(2016) Molecular dynamics simulations of membrane proteins and their interactions: from nanoscale to mesoscale.', 'Curr.', 'Opin.', 'Struct.', 'Biol., 40, 8–16.', 'Cheng,L.S.', 'et al.', '(2008) Ensemble-based virtual screening reveals potential\\n novel antiviral compounds for avian influenza neuraminidase.', 'J. Med.', 'Chem., 51, 3878–3894.', 'd’Errico,M.', 'et al.', '(2018) Automatic topography of high-dimensional data sets\\n by non-parametric density peak clustering.', 'arXiv:1802.10549v1 [stat.ML]\\n De Paris,R.', 'et al.', '(2015) Clustering molecular dynamics trajectories for optimizing docking experiments.', 'Comput.', 'Intell.', 'Neurosci., 2015, 1–9.', 'de Souza,V.C.', 'et al.', '(2017) Clustering algorithms applied on analysis of protein molecular dynamics.', 'In: 2017 IEEE Latin American Conference on\\n Computational Intelligence (LA-CCI).', 'Arequipa, 2017, pp.', '1–6.', 'https://doi.', 'org/10.1109/LA-CCI.2017.8285695\\n De Vivo,M.', 'et al.', '(2016) Role of molecular dynamics and related methods in\\n drug discovery.', 'J. Med.', 'Chem., 59, 4035–4061.', 'Degiacomi,M.T.', 'and Dal Peraro,M.', '(2013) Macromolecular symmetric assembly prediction using swarm intelligence dynamic modeling.', 'Structure,\\n 21, 1097–1106.', 'Degiacomi,M.T.', 'et al.', '(2013) Molecular assembly of the aerolysin pore reveals\\n a swirling membrane-insertion mechanism.', 'Nat.', 'Chem.', 'Biol., 9, 623–629.', 'Doerr,S.', 'et al.', '(2016) HTMD: high-throughput molecular dynamics for molecular discovery.', 'J. Chem.', 'Theory Comput., 12, 1845–1852.', 'Du,M.', 'et al.', '(2016) Study on density peaks clustering based on k-nearest neighbors and principal component analysis.', 'Knowl.', 'Based Syst., 99, 135–145.', 'Ester,M.', 'et al.', '(1996) A density-based algorithm for discovering clusters a\\n density-based algorithm for discovering clusters in large spatial databases\\n with noise.', 'In: Proceedings of the Second International Conference on\\n Knowledge Discovery and Data Mining, KDD’96, pp.', '226–231.', 'AAAI\\n Press, Portland, OR, USA.', 'Fonti,G.', 'et al.', '(2019) KAP1 is an antiparallel dimer with a natively functional\\n asymmetry.', 'Life Science Alliance, 2(4), e201900349.', 'Frank,J.', '(2018) New opportunities created by single-particle cryo-EM: the\\n mapping of conformational space.', 'Biochemistry, 57, 888–888.', 'Fra¨nti,P., and Sieranoja,S.', '(2018) K-means properties on six clustering benchmark datasets.', 'Applied Intelligence, 48, 4743–4759.', '10.1007/s10489-\\n 018-1238-7\\n Fig.', '5.', 'Classification and frequency estimation of APP dimerization motifs in the\\n plasma membrane.', '(a) The two known dimerization motifs of the transmembrane\\n helix of APP.', 'Membrane is depicted with gray spheres, the G700G704G708 motif\\n with teal spheres and the G709A713 motif with gold spheres.', '(b) The black line\\n depicts the probability density distribution of the trajectory along the first principal\\n component, which covers 77% of variability.', 'Obtained clusters are shown in colors\\n underneath, with their centers as black stars.', 'Below the clusters is the distribution of\\n core cardinalities, with the corresponding color bar below it.', '(c) The cluster of\\n bound dimers of (b) was extracted and the pair-wise distances between the backbone atoms of each motifs in both helices was computed for every frame.', 'The plot\\n shows the first two principal components of the resulting dataset (with eigenvalues\\n of 0.66 and 0.28, respectively), where each point is color-coded according to the\\n cardinality of its core.', 'Stars represent cluster centers.', '(d) The clusters obtained from\\n the aforementioned dataset of bound conformations.', '(e) Renders of the unbound\\n and dimerized APP cluster centers and their respective frequency after outlier removal.', 'Dimerized APP centers shown correspond to the dark blue, dark brown and\\n dark green clusters.', 'Sidechains have been hidden for better visualization\\n CLoNe: automated clustering of biomolecular structural ensembles 7\\n Downloaded from https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa742/5895303 by guest on 19 March 2021\\n Fu,L.', 'and Medico,E.', '(2007) FLAME, a novel fuzzy clustering method for the\\n analysis of DNA microarray data.', 'BMC Bioinformatics, 8, 3.', 'Gionis,A.', 'et al.', '(2007) Clustering aggregation.', 'ACM Trans.', 'Knowl.', 'Discov.', 'Data, 1, 4.', 'Hamelberg,D.', 'et al.', '(2004) Accelerated molecular dynamics: a promising and\\n efficient simulation method for biomolecules.', 'J. Chem.', 'Phys., 120,\\n 11919–11929.', 'Horn,J.R.', 'and Shoichet,B.K.', '(2004) Allosteric inhibition through core disruption.', 'J. Mol.', 'Biol, 336, 1283–1291.', 'Humphrey,W.', 'et al.', '(1996) VMD: visual molecular dynamics.', 'J. Mol.', 'Graph,\\n 14, 33–38.', 'Husic,B.E.', 'and Pande,V.S.', '(2018) Markov state models: from an art to a science.', 'J.', 'Am.', 'Chem.', 'Soc., 140, 2386–2396.', 'Husic,B.E.', 'and Pande,V.S.', '(2017) Ward clustering improves cross-validated Markov\\n state models of protein folding.J.', 'Chem.', 'Theory Comput., 13, 963–967.', 'Jain,A.K.', '(2010) Data clustering: 50 years beyond K-means.', 'Pattern Recognit.', 'Lett., 31, 651–666.', 'Jones,E.', 'et al.', '(2001) SciPy: open source scientific tools for Python.', '10.1038/s41592-019-0686-2.', 'Kokh,D.B.', 'et al.', '(2011) Receptor flexibility in small-molecule docking calculations.', 'Wiley Interdiscip.', 'Rev.', 'Comput.', 'Mol.', 'Sci., 1, 298–314.', 'Le Guilloux,V.', 'et al.', '(2009) Fpocket: an open source platform for ligand\\n pocket detection.', 'BMC Bioinformatics, 10, 168.', 'Liang,Z.', 'and Chen,P.', '(2016) Delta-density based clustering with a\\n divide-and-conquer strategy: 3DC clustering.', 'Pattern Recognit.', 'Lett., 73, 52–59.', 'Lohman,D.C.', 'et al.', '(2019) An isoprene lipid-binding protein promotes eukaryotic coenzyme Q biosynthesis.', 'Mol.', 'Cell, 73, 763–774.e10.', 'Lohman,D.C.', 'et al.', '(2014) Mitochondrial COQ9 is a lipid-binding protein\\n that associates with COQ7 to enable coenzyme Q biosynthesis.', 'Proc.', 'Natl.', 'Acad.', 'Sci.', 'USA, 111, E4697–E4705.', 'Malhotra,S.', 'et al.', '(2019) Modelling structures in cryo-EM maps.', 'Curr.', 'Opin.', 'Struct.', 'Biol., 58, 105–114.', 'McGibbon,R.T.', 'et al.', '(2015) MDTraj: a modern open library for the analysis\\n of molecular dynamics trajectories.', 'Biophys.', 'J., 109, 1528–1532.', 'McKiernan,K.', 'A. et al.. (2017) Modeling the mechanism of CLN025 beta--\\n hairpin formation.', 'The Journal of Chemical Physics, 147, 104107.', 'Mehmood,R.', 'et al.', '(2016) Clustering by fast search and find of density peaks\\n via heat diffusion.', 'Neurocomputing, 208, 210–217.', 'Mu,Y.', 'et al.', '(2004) Energy landscape of a small peptide revealed by dihedral\\n angle principal component analysis.', 'Proteins Struct.', 'Funct.', 'Bioinform., 58,\\n 45–52.', 'Naritomi,Y.', 'and Fuchigami,S.', '(2011) Slow dynamics in protein fluctuations\\n revealed by time-structure based independent component analysis: the case\\n of domain motions.', 'J. Chem.', 'Phys., 134, 065101.', 'Noe´,F.', 'et al.', '(2019) Boltzmann generators: sampling equilibrium states of\\n many-body systems with deep learning.', 'Science, 365, eaaw1147.', 'Oleinikovas,V.', 'et al.', '(2016) Understanding cryptic pocket formation in protein targets by enhanced sampling simulations.', 'J.', 'Am.', 'Chem.', 'Soc., 138,\\n 14257–14263.', 'Paris,R.D.', 'et al.', '(2015) An effective approach for clustering InhA molecular\\n dynamics trajectory using substrate-binding cavity features.', 'PLoS One, 10,\\n e0133172.', 'Pedregosa,F.', 'et al.', '(2011) Scikit-learn: machine learning in Python.', 'J. Mach.', 'Learn.', 'Res., 12, 2825–2830.', 'Peng,J.', 'et al.', '(2018) Clustering algorithms to analyze molecular dynamics\\n simulation trajectories for complex chemical and biological systems.', 'Chin.', 'J. Chem.', 'Phys., 31, 404–420.', 'Pe´rez-Herna´ndez,G.', 'et al.', '(2013) Identification of slow molecular order\\n parameters for Markov model construction.', 'J. Chem.', 'Phys., 139, 015102.', 'Rodriguez,A.', 'et al.', '(2018) Computing the free energy without collective variables.', 'J. Chem.', 'Theory Comput., 14, 1206–1215.', 'Rodriguez,A.', 'and Laio,A.', '(2014) Clustering by fast search and find of density\\n peaks.', 'Science, 344, 1492–1496.', 'Salmaso,V., and Moro,S.', '(2018) Bridging molecular docking to molecular dynamics in exploring ligand-protein recognition process: an overview.', 'Front.', 'Pharmacol, 9, 923.', 'Seabold,S.', 'and Perktold,J.', '(2010) Statsmodels: econometric and statistical\\n modeling with python.', 'In: 9th Python in Science Conference.', 'https://www.', 'statsmodels.org/devel/#citation.', 'Shao,J.', 'et al.', '(2007) Clustering molecular dynamics trajectories: 1.', 'Characterizing the performance of different clustering algorithms.', 'J. Chem.', 'Theory Comput., 3, 2312–2334.', 'Shirts,M.', 'and Pande,V.S.', '(2000) Screen savers of the World Unite!', 'Science,\\n 290, 1903–1904.', 'Sultan,M.M.', 'et al.', '(2018) Transferable neural networks for enhanced sampling of protein dynamics.', 'J. Chem.', 'Theory Comput., 14, 1887–1894.', 'Tamo`,G.E.', 'et al.', '(2015) The importance of dynamics in integrative modeling\\n of supramolecular assemblies.', 'Curr.', 'Opin.', 'Struct.', 'Biol., 31, 28–34.', 'Vahl Quevedo,C.', 'et al.', '(2014) A strategic solution to optimize molecular\\n docking simulations using Fully-Flexible Receptor models.', 'Expert Syst.', 'Appl., 41, 7608–7620.', 'Wang,W.', 'et al.', '(2018) Constructing Markov state models to elucidate the\\n functional conformational changes of complex biomolecules.', 'Wiley\\n Interdiscip.', 'Rev.', 'Comput.', 'Mol.', 'Sci., 8, e1343.', 'Wang,X.-F. and Xu,Y.', '(2017) Fast clustering using adaptive density peak detection.', 'Stat.', 'Methods Med.', 'Res., 26, 2800–2811.', 'Ward,J.H.', '(1963) Hierarchical grouping to optimize an objective function.', 'J.', 'Am.', 'Stat.', 'Assoc., 58, 236–244.', 'Wassenaar,T.A.', 'et al.', '(2015) High-throughput simulations of dimer and trimer assembly of membrane proteins.', 'the DAFT Approach.', 'J. Chem.', 'Theory\\n Comput., 11, 2278–2291.', 'Wolf,A.', 'and Kirschner,K.N.', '(2013) Principal component and clustering analysis on molecular dynamics data of the ribosomal L11\\x0523S subdomain.', 'J.\\n Mol.', 'Model., 19, 539–549.', 'Xie,J.', 'et al.', '(2016) Robust clustering by detecting density peaks and assigning\\n points based on fuzzy weighted K-nearest neighbors.', 'Inf.', 'Sci., 354, 19–40.', 'Zhang,W.', 'and Li,J.', '(2015) Extended fast search clustering algorithm: widely\\n density clusters, no density peaks.', 'arXiv:1505.05610 [cs.DS].', '8 S.Tra¨ ger et al.', 'Downloaded from https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa742/5895303 by guest on 19 March 2021'][' Structural bioinformatics\\n CLoNe: automated clustering based on local density\\n neighborhoods for application to biomolecular structural\\n ensembles\\n Sylvain Tra¨ ger1,2, Giorgio Tamo` 1,2, Deniz Aydin1,2, Giulia Fonti1,2,\\n Martina Audagnotto1,2 and Matteo Dal Peraro1,2,*\\n 1\\n Institute of Bioengineering, School of Life Sciences, Ecole Polytechnique Fe´ de´ rale de Lausanne, Lausanne 1025, Switzerland and 2\\n Swiss Institute of Bioinformatics, Lausanne 1015, Switzerland\\n *To whom correspondence should be addressed.', 'Associate Editor: Lenore Cowen\\n Received on January 8, 2020; revised on July 14, 2020; editorial decision on August 12, 2020; accepted on August 18, 2020\\n Abstract\\n Motivation: Proteins are intrinsically dynamic entities.', 'Flexibility sampling methods, such as molecular dynamics or\\n those arising from integrative modeling strategies, are now commonplace and enable the study of molecular conformational landscapes in many contexts.', 'Resulting structural ensembles increase in size as technological and algorithmic advancements take place, making their analysis increasingly demanding.', 'In this regard, cluster analysis\\n remains a go-to approach for their classification.', 'However, many state-of-the-art algorithms are restricted to specific\\n cluster properties.', 'Combined with tedious parameter fine-tuning, cluster analysis of protein structural ensembles\\n suffers from the lack of a generally applicable and easy to use clustering scheme.', 'Results: We present CLoNe, an original Python-based clustering scheme that builds on the Density Peaks algorithm\\n of Rodriguez and Laio.', 'CLoNe relies on a probabilistic analysis of local density distributions derived from nearest\\n neighbors to find relevant clusters regardless of cluster shape, size, distribution and amount.', 'We show its capabilities on many toy datasets with properties otherwise dividing state-of-the-art approaches and improves on the original algorithm in key aspects.', 'Applied to structural ensembles, CLoNe was able to extract meaningful conformations from membrane binding events and ligand-binding pocket opening as well as identify dominant dimerization\\n motifs or inter-domain organization.', 'CLoNe additionally saves clusters as individual trajectories for further analysis\\n and provides scripts for automated use with molecular visualization software.', 'Availability and implementation: www.epfl.ch/labs/lbm/resources, github.com/LBM-EPFL/CLoNe.', 'Contact: matteo.dalperaro@epfl.ch\\n Supplementary information: Supplementary data are available at Bioinformatics online.', '1 Introduction\\n The perception of molecular structures, especially proteins, is gradually shifting from the concept of one single and rigid structure to\\n the idea that biomolecules natively exhibit a continuum of states\\n (Frank, 2018).', 'Protein folding, post-translational modifications\\n (Audagnotto and Dal Peraro, 2017), binding to other molecules or\\n their involvement in catalytic events result in vast and complex conformational landscapes.', 'Molecular dynamics (MD), thanks to progress in both its technological and algorithmic aspects, allows for\\n the simulation of key biomolecular events.', 'Their observability, however, tends to be limited by currently accessible timescales.', 'Researchers consistently come up with innovative protocols to push\\n this limit further (Barducci et al., 2011; Bussi, 2014; Chavent et al.,\\n 2016; Doerr et al., 2016; Hamelberg et al., 2004; Noe´ et al., 2019;\\n Shirts and Pande, 2000; Sultan et al., 2018; Wassenaar et al., 2015)\\n granting us with the ability to capture protein folding as well as protein–protein, protein–membrane and protein–ligand interactions\\n (Audagnotto et al., 2016; De Vivo et al., 2016; McKiernan et al.,\\n 2017; Oleinikovas et al., 2016).', 'State-of-the-art protocols for smallmolecule docking (Amaro et al., 2018; Kokh et al., 2011; Vahl\\n Quevedo et al., 2014), protein–protein docking and integrative\\n modeling strategies, in general, have shifted toward the integration\\n of dynamics in some form as well (Abriata and Dal Peraro, 2020;\\n Malhotra et al., 2019; Tamo` et al., 2015).', 'All of the aforementioned\\n aspects advocate dynamics as a cornerstone of modern structural\\n biology and push the need for efficient tools to extract functional insight from structural ensembles in general.', 'VC The Author(s) 2020.', 'Published by Oxford University Press.', '1\\n This is an Open Access article distributed under the terms of the Creative Commons Attribution Non-Commercial License (http://creativecommons.org/licenses/by-nc/4.0/),\\n which permits non-commercial re-use, distribution, and reproduction in any medium, provided the original work is properly cited.', 'For commercial re-use, please contact\\n journals.permissions@oup.com\\n Bioinformatics, 2020, 1–8\\n doi: 10.1093/bioinformatics/btaa742\\n Advance Access Publication Date: 21 August 2020\\n Original Paper\\n Downloaded from https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa742/5895303 by guest on 19 March 2021\\n However, these advances come at a price.', 'The sheer size, the intrinsic complexity and redundancy of structural ensembles makes\\n their successful analysis and computational integration non-trivial.', 'Coarse-graining tools such as cluster analysis effectively reduce simulations of thousands of conformations to few key biological states\\n and hence constitute a go-to approach with countless applications to\\n date [(Cheng et al., 2008; De Paris et al., 2015; de Souza et al.,\\n 2017), reviewed in (Peng et al., 2018; Shao et al., 2007)].', 'Such states\\n may serve as basis of Markov state models (Husic and Pande, 2018;\\n Wang et al., 2018).', 'To our knowledge, however, an algorithm able\\n to cluster data efficiently irrespective of their properties is still missing.', 'Indeed, different cluster shapes, sizes and densities usually dictate which clustering approach is best suited for a given task.', 'The\\n most known and widely used scheme is probably that of k-means\\n and its many variations (Jain, 2010).', 'This center-based clustering\\n scheme, however, suffers from unequal results due to random initializations, the a priori setting of the number of clusters and its limitation to spherical clusters.', 'Alternatively, hierarchical schemes such as\\n the Ward-linkage agglomerative algorithm (Ward, 1963) do not require pre-setting the number of clusters and are popular for building\\n Markov state models (Beauchamp et al., 2012; Husic and Pande,\\n 2017; Paris et al., 2015).', 'They are however sensitive to noise and\\n outliers and may suffer from non-spherical clusters (Peng et al.,\\n 2018).', 'Conversely, DBSCAN (Ester et al., 1996) is able to manage\\n clusters regardless of shape by utilizing density differences between\\n clusters and noise.', 'However, setting its parameter is not trivial and\\n its optimal value may not be unique throughout the dataset when\\n clusters of largely different densities are present.', 'This limitation is at\\n the core of OPTICS, which can be seen as an extension of DBSCAN\\n (Ankerst et al., 1999), although not strictly advertised as a clustering\\n algorithm.', 'Defining metastable states of proteins is non-trivial due to the\\n large and often redundant number of internal degrees of freedom,\\n yielding sampled conformational spaces with local minima often devoid of biological significance.', 'We can make the assumption that,\\n given enough sampling and a choice of relevant features, metastable\\n states would lie in regions or clusters of high density, which would\\n be separated by valleys of different density levels that would correspond to transitional states.', 'Furthermore, no assumption can be\\n made on the shape or relative densities of clusters, which would depend on both conformational sampling and target system.', 'Rodriguez and Laio (2014) designed the Density Peaks (DP) algorithm aimed at clustering regardless of shape and dimensionality.', 'Their algorithm generated significant interest thanks to their clever\\n definition of cluster centers, which states that a cluster center should\\n display a higher density (q) than its neighbors and a high distance to\\n another point of higher density (d).', 'DP takes a single input parameter, which relates to a cutoff distance for the computation of q.', 'However, it requires the user to specify thresholds for both q and d\\n mid-computation in order to select the cluster centers, which prevents a fully automated clustering process.', 'DP has since been\\n improved by the inclusion of k nearest neighbors (kNN) (Du et al.,\\n 2016; Xie et al., 2016; Zhang and Li, 2015) or heat diffusion\\n (Mehmood et al., 2016) for a more robust estimation of q, which\\n allows for a better handling of cases where clusters have significantly different densities.', 'These improvements still require user intervention for selecting cluster centers.', 'Conversely, Wang and Xu (2017)\\n built on DP to automatically select cluster centers based on maximizing an average silhouette index, although other input parameters\\n are required instead.', 'Liang and Cheng coupled principles from\\n DBSCAN with a divide-and-conquer approach to recursively and\\n automatically select cluster centers (Liang and Chen, 2016).', 'Recently, d’Errico et al.', 'coupled DP with a non-parametric density\\n estimator (Rodriguez et al., 2018), yielding Density Peaks Advanced\\n (DPA; d’Errico et al., 2018).', 'While exhibiting impressive robustness\\n to a variety of cluster shapes, densities and to outliers, DPA still suffers from a few issues.', 'We found that it performed worse than the\\n original on some typical benchmark datasets, and requires a sensitive albeit unique input parameter.', 'Moreover, both DP and DPA exhibit an inconsistent outlier removal procedure.', 'These drawbacks\\n may prove crucial when targeting structural biological data, where\\n regions at lower effective density may have equal or even increased\\n significance than others at higher densities.', 'The complexity of biological structures leads to numerous unique yet equally relevant\\n choices of features, each with their own topology.', 'The analysis of\\n such datasets is greatly hindered by sensitivity to input parameters,\\n which implies that tedious fine-tuning steps have to be undertaken.', 'Here, we introduce an approach to remedy these drawbacks,\\n enabling a facilitated analysis of complex real-world datasets from\\n structural biology.', 'Our approach builds on the original DP algorithm by introducing a fragmenting of the data into specific density\\n distributions.', 'In essence, the local densities of each point are computed using nearest neighbors and a Gaussian kernel and points\\n associated with local density maxima are identified as putative cluster centers.', 'To increase robustness to non-spherical cluster shapes,\\n clusters are merged using the Bhattacaryaa coefficient\\n (Bhattacharyya, 1943) by comparing density distributions derived\\n from putative cluster cores and boundaries.', 'Finally, outliers from\\n impromptu noise fluctuations are removed by means of a Bayes classifier.', 'This, to the best of our knowledge, constitute an original contribution to the density peaks algorithm.', 'Termed Clustering based\\n on Local density Neighborhoods (CLoNe), our approach relies on a\\n single input parameter that is both robust and intuitive to set.', 'We\\n test it on many typical benchmark datasets and against state-of-theart clustering schemes.', 'The local focus of CLoNe allows for the detection of biological states of smaller frequency while its ease of use\\n allows the researcher to focus on choosing relevant biological features for pre-processing or analyzing their structural ensemble without being hindered by algorithmic limitations.', 'Furthermore, CLoNe\\n outputs useful molecular visualization scripts for the validation of\\n cluster relevance in the target biological context (Supplementary Fig.', 'S1).', 'We apply CLoNe on a range of structural datasets from MD\\n simulations or integrative modeling studies, each time detailing the\\n feature selection process and which information can be extracted\\n from the results.', 'Our examples cover previously published studies\\n on protein–membrane interactions, internal structural rearrangements of disordered proteins, cryptic allosteric pocket formation and\\n transmembrane dimerization motifs, and highlight the broad advantages of CLoNe for the analysis of molecular structural ensembles.', '2 Materials and methods\\n An overview and basic usage of CLoNe is available in\\n Supplementary Figure S1, on GitHub and the webpage of our laboratory (see Abstract section).', 'We created a synthetic dataset containing clusters of significantly different densities and various shapes\\n in order to showcase the procedure behind our approach.', 'CLoNe\\n starts by finding the k nearest neighbors of each point in a dataset X\\n of N points using k-nearest neighbors (kNN) (Pedregosa et al.,\\n 2011), yielding a neighbor matrix M where each row i contains all\\n the neighbors j of point i in increasing order of Euclidean distance.', 'To account for significant density differences between clusters, we\\n initially assume that all points are cluster centers.', 'In a first step, we\\n estimate the local density q of each point i using a Gaussian kernel:\\n qi ¼ X\\n jkNNi\\n e\\n  Mi;j\\n dc\\n  2\\n ¼ X\\n jkNNi\\n qij (1)\\n where kNNi is the set of nearest neighbors of i in increasing order of\\n distance and dc is a cutoff distance defined as to be superior to a\\n user-defined percentage pdc (the single input parameter of CLoNe)\\n of all distances within M, similar to the original DP algorithm\\n (Rodriguez and Laio, 2014).', 'We define the core of putative cluster i\\n as the set of neighbors that contribute to qi at least as much of the\\n j1 previous neighbors on average:\\n corei ¼ j  kNNijqij\\n 1\\n pdcN\\n Xj1\\n k¼0 qik \\x04 \\x05 (2)\\n We show in Figure 1a the cardinality (number of elements) of\\n the core of each point in our synthetic dataset.', 'As expected, this\\n 2 S.Tra¨ ger et al.', 'Downloaded from https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa742/5895303 by guest on 19 March 2021\\n number is higher for points closer to real cluster centers and lower\\n for points laying on the outskirt of clusters.', 'The visualization of core\\n cardinalities is an efficient way to observe the underlying topology\\n of the dataset.', 'In order to identify if i is a genuine candidate for cluster center, we identify its first neighbor j of higher density.', 'If neighbor j belongs to corei, then neighbor j is a better candidate for\\n cluster center in this region than i. Conversely, i is a genuine candidate for cluster center if j is not in the core of i.', 'Cluster assignation\\n is done in a single step by assigning a point to the same cluster as its\\n nearest point of higher density, in order of decreasing local density,\\n similar to the original DP approach.', 'The results at this stage of\\n CLoNe are shown in Figure 1a.', 'One of the drawbacks of the DP algorithm is its limited ability to\\n deal with clusters with more than one peak or those with an elongated region of similar density, such as the noisy circles benchmark\\n dataset (Supplementary Fig.', 'S2).', 'This is true at this stage of CLoNe\\n as well, as the Gaussian kernel in (1) is biased toward determining\\n cores with spherical shapes (Fig.', '1b).', 'We can make the assumptions\\n that if two existing clusters A and B should be merged, then the\\n density from one core to the other should be relatively constant.', 'This can be estimated by looking at the core cardinality distribution\\n of the points belonging to the core of both clusters as well as that of\\n the points from the boundary between them (Fig.', '1b, left), which\\n can be defined as:\\n boundaryAB ¼ i 2 A; j 2 Bjd ið Þ ; j < dcg \\x06 (3)\\n Then, we define the following probability density function for\\n the ensemble of points belonging to either cluster cores or the\\n boundary from (3):\\n Ps ¼ KDE corei ð Þ f g ; i 2 S (4)\\n where S denotes one of the aforementioned ensembles and #corei the\\n core cardinality of point i and KDE refers to the probability density\\n function estimated by unimodal Gaussian kernel density estimation.', 'Similarity between probability distributions can be measured using\\n the Bhattacharyya coefficient (BC) (Bhattacharyya, 1943), which is\\n bound between 0 and 1.', 'Thus, the formula to compute the BC between the core of cluster A with the boundary from (3) becomes:\\n BCA ¼ ffiffiffiffiffiffiffiffiffiffiffiffiffiffiffiffiffiffiffiffiffiffiffiffi\\n PboundPcoreA\\n p (5)\\n We take the decision of merging Clusters A and B if the mean of\\n their respective BC with their boundary is above a threshold sBC:\\n B Að Þ ; B ¼ 1\\n 2 ð Þ BCA þ BCB > sBC (6)\\n where PcoreB is obtained as for cluster A in (5) and sBC was chosen to\\n be the 65th percentile to limit uncertainty and based on benchmarks\\n (Fig.', '1 and Supplementary Fig.', 'S2).', 'Taking the mean of both coefficients prevents the merging of a cluster whose probability density is\\n similar to that of its boundary with a cluster of significantly higher\\n density.', 'This enabled us to identify clusters that can hardly be\\n defined with a single density peak, such as uniform density over\\n non-spherical shapes (Figs 1b and 2a).', 'The point with highest q is\\n chosen as the new cluster center.', 'If pdc is chosen too small, clusters\\n may be split into sub-clusters.', 'Within our approach, these subclusters are likely to be merged into clusters matching the original\\n topology, expanding the range of acceptable values for pdc and making it an input parameter less sensitive than the one of DPA\\n (Supplementary Fig.', 'S3).', 'The second drawback of using the Gaussian kernel in (1) is that\\n it may falsely identify impromptu local noise fluctuations as cluster\\n centers (Fig.', '1b).', 'To remedy this, we define two probability density\\n functions.', 'Mcores is the probability density of the local density of all\\n points belonging to any cluster core as per (2):\\n Mcores ¼ KDEð qx f g ; x 2 corei 8i 2 C Þ (7)\\n where C is the set of cluster centers remaining after the previous\\n merging step.', 'Moutliers is the equivalent function for all points identified as outliers:\\n Moutliers ¼ KDEð qx; x 2 X if qx < f \\x04 qcenterx\\n \\x06 \\x08Þ (8)\\n where qcenterx is the local density of the center of the cluster x\\n belongs to and f an arbitrary fraction chosen to be 0.1 by default\\n and used throughout this article.', 'To determine if an identified cluster is more probable to be derived from noise than to be a genuine\\n cluster, we use a Bayesian classifier.', 'For each cluster core ci, we derive the following posterior probabilities using Bayes’ theorem:\\n p Yð Þ¼ jci\\n p Yð Þ pðcijYÞ\\n pðciÞ ; Y 2 f g outliers; cores (9)\\n where the prior probabilities are defined as follows:\\n Fig.', '1.', 'Clustering based on local density neighborhoods.', 'We created a synthetic\\n dataset that contains four croissant-shaped clusters of 500 elements each with different scaling as well as a cluster in the shape of a cross with 4000 elements and added\\n Gaussian noise.', '(a) On the left, the cardinality of the core of each point in the dataset.', 'On the right, the clusters obtained after the first stage of clustering.', '(b) The average Bhattacaryya coefficient B(A,B) between two Clusters A and B is shown.', 'The\\n upper left plots show an example of two clusters being merged corresponding to the\\n upper square in panel (a).', 'The lower left plots show an example of two clusters that\\n will not be merged corresponding to the lower square in panel (a).', 'Points belonging\\n to the core of each cluster are shown in black, regular points in shades of gray and\\n points belonging to the boundary in red.', 'The plot on the right shows the clustering\\n after merging clusters.', '(c) CLoNe uses a Bayesian classifier to decide if a cluster is\\n genuine or arises from noise fluctuations.', 'On the left, the corresponding probability\\n density functions of points belonging to any cluster cores (black), noise (blue), or to\\n individual cluster cores (light blue).', 'The range of local density of clusters classified\\n as noise fluctuations are shown on the secondary y-axis (dark blue).', 'The final clustering result is shown on the right, with identified outliers shown as black crosses\\n CLoNe: automated clustering of biomolecular structural ensembles 3\\n Downloaded from https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa742/5895303 by guest on 19 March 2021\\n p Yð Þ ¼ jYj\\n N ; Y 2 f g outliers; cores (10)\\n where jYj denotes the cardinality of the corresponding class.', 'The\\n likelihoods pðcijYÞ can be computed by evaluating the previously\\n defined probability distributions (7) and (8) at ci.', 'Disregarding the\\n evidence pðciÞ common for both outliers and cores classes, we thus\\n obtain the following Bayesian classifier:\\n y^i ¼ argmax Y2f g outliers; cores\\n p Yð ÞY\\n x2ci\\n MYðqxÞ (11)\\n This classifier enabled us to remove all cluster centers arising\\n from noise fluctuations.', 'Combined with the previous merging step,\\n the clustering is now complete (Fig.', '1c).', 'The software has been written using Python3.7.', 'When used on\\n structural biological data, CLoNe outputs cluster centers as separate\\n PDB files, individual clusters as XTC trajectories and Tcl scripts for\\n automated loading within the visualization software VMD (Visual\\n Molecular Dynamics, (Humphrey et al., 1996)).', 'The loading of MD\\n trajectories as well as the saving of cluster centers and cluster subtrajectories is done through the MDTraj package (McGibbon et al.,\\n 2015).', 'We use the scipy (Jones et al., 2001), scikit-learn (Pedregosa\\n et al., 2011) and Statsmodels (Seabold and Perktold, 2010) packages\\n for many operations described in the previous subsection and to\\n compare CLoNe to other clustering algorithms.', '3 Results\\n 3.1 Automatic cluster center determination from local\\n density neighborhood analysis\\n We applied CLoNe to a large set of common benchmark datasets\\n from various sources (Chang and Yeung, 2008; d’Errico et al., 2018;\\n Fra¨nti and Sieranoja, 2018; Fu and Medico, 2007; Gionis et al.,\\n 2007; Pedregosa et al., 2011), covering different key properties of\\n clusters, such as non-spherical shapes, anisotropy, as well as significant size and density differences, all of which can be expected from\\n real-world datasets from structural biology.', 'In the previous section,\\n we detailed how CLoNe automatically detects cluster centers, accurately merges clusters and removes outliers, succeeding in cases where\\n previous iterations of DP did not (Fig.', '2a and Supplementary Fig.', 'S2).', 'Similarly to the original DP algorithm (Rodriguez and Laio,\\n 2014), CLoNe requires a single input parameter pdc, which relates\\n to a cutoff distance used in the estimation of local densities (see\\n Section 2).', 'In general, pdc takes a value in a small range and is\\n intuitive to set.', 'For instance, with pdc values of 1 or 2 local densities\\n will be estimated considering neighborhoods small enough to identify individual spiral branches as clusters (Fig.', '2b).', 'For higher values\\n of pdc, the scale of the Gaussian kernel in Equation (1) will increase\\n and merge individual branches into whole spirals, allowing the study\\n of multiple hierarchies intuitively (Fig.', '2c).', 'Other than clusters with\\n non-spherical shapes, CLoNe identifies successfully the numerous\\n Gaussian clusters of the A3 dataset (Fig.', '2d).', 'Some degree of overlapping in real-world datasets is to be expected.', 'The S4 dataset contains 15 highly overlapping Gaussian clusters of varying densities\\n and shapes but equal size.', 'As with the A3 dataset, CLoNe does not\\n perform unnecessary merging with nearby clusters (Fig.', '2e) and is\\n robust to large amounts of outliers on top of clusters with significantly different densities (Fig.', '2f).', 'This general applicability of\\n CLoNe coupled with a single, robust and easy to set input parameter\\n (Supplementary Fig.', 'S3) is unique among the commonly used clustering algorithms found in the Scikit-learn package (Pedregosa et al.,\\n 2011).', 'In fact, CLoNe is among the fastest algorithms from that\\n package in addition to both DP and DPA and the most accurate on\\n the available benchmark cases (Supplementary Fig.', 'S2).', 'The only\\n other algorithm succeeding on all benchmark cases is OPTICS\\n (Ankerst et al., 1999), which runs slightly slower than CLoNe on\\n these datasets and tends to classify too many points as outliers.', 'Similar to the original implementation of DP, CLoNe is applicable\\n to high dimensionality datasets as well (Supplementary Table S1\\n and Supplementary Fig.', 'S4).', 'One of the principal aims of this work is to offer a clustering algorithm able to classify unlabeled biological structural ensembles\\n into relevant states associated with their function and mechanism of\\n action.', 'We have applied CLoNe to real-world structural biology\\n data reporting on the dynamic conformational space of a protein\\n that associates with its specific biological membrane, cryptic allosteric pocket opening and dimerization of transmembrane proteins.', '3.2 Determining relevant states within protein\\n conformational ensembles\\n COQ9 is a lipid-binding protein associated with the biosynthesis of\\n coenzyme Q (CoQ), a redox-active lipid that is essential for cellular\\n respiration (Lohman et al., 2014).', 'Recently, coarse-grained molecular dynamics (CG-MD) simulations and liposome co-flotation assays\\n were used together to reveal that COQ9 accesses membranes in a\\n multi-step fashion through a distinct, C-terminal amphipathic helix\\n (a10) (Lohman et al., 2019).', 'In these simulations, COQ9 first diffused in the aqueous environment, then underwent various\\n Fig.', '2.', 'Local neighborhood density analysis for automated center and cluster determination.', 'For each panel, the cardinality of the core of each point as detailed in the Section 2\\n is shown on the left.', 'The resulting clusters are shown on the right, with clusters in different colors and identified outliers as black crosses.', '(a) Noisy circles dataset.', '(b and c)\\n Four instances of spiral dataset with different values of the input parameter yielding different yet valid clusters.', '(d) The a3 dataset containing 50 Gaussian clusters.', '(e) The s4\\n dataset with highly overlapping Gaussian clusters, some with anisotropic distributions.', '(f) A synthetic dataset with clusters of significantly different sizes and densities taken\\n from Density Peaks Advanced\\n 4 S.Tra¨ ger et al.', 'Downloaded from https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa742/5895303 by guest on 19 March 2021\\n conformational changes upon membrane binding (Lohman et al.,\\n 2019).', 'We applied CLoNe to the CG-MD trajectory used in the latter study and sought to identify the main binding events pertaining\\n to the protein itself.', 'To this end, we extracted features characterizing both its movements in the aqueous environment through monitoring its distance to the membrane as well as key conformational\\n changes based on the angle between the unique a10 helix of COQ9\\n and its globular domain (Fig.', '3a, left).', 'Using these two features,\\n CLoNe outputs three clusters, each of which seem to follow\\n Gaussian distributions (Fig.', '3b).', 'One cluster regroups all conformations that correspond to diffusion movements in the aqueous environment, while the other two highlight the membrane association of\\n a10 first followed by the globular domain as a converging step\\n (Fig.', '3a), thus its higher density (Fig.', '3b).', 'Similar results can be\\n obtained hypothesis-free by using raw atomic spatial coordinates\\n (Supplementary Fig.', 'S4).', 'The Gaussian distribution of structural\\n clusters has also been observed in a recent study from our group\\n involving the KAP1 protein, where CLoNe was also successfully\\n applied (Supplementary Fig.', 'S5) (Fonti et al., 2019).', '3.2 Isolating sub-ensembles of relevant conformations\\n for ligand–target interactions\\n In recent years, small-molecule docking software is no stranger to\\n dynamics, taking into account ensembles of ligand conformations\\n (Amaro et al., 2018) or receptor flexibility (Kokh et al., 2011;\\n Salmaso and Moro, 2018; Vahl Quevedo et al., 2014).', 'A recent\\n study highlighted a novel replica exchange-based MD protocol combined with benzene probes, where each replica harbors a different\\n scaling of water–protein interactions (Oleinikovas et al., 2016).', 'Using this method, the authors could observe the opening of cryptic\\n allosteric pockets in several systems, including that of the TEM1 blactamase, which plays a critical role in antibiotic resistance (Horn\\n and Shoichet, 2004).', 'The simulations were started from the apo\\n crystal structure with a closed allosteric pocket (Fig.', '4a).', 'Out of the\\n eight replicas of the simulation, we chose three with neutral (first),\\n medium (fourth) and highest (last) scaling factors as a tradeoff between maximizing the sampled conformational space and limiting\\n redundancy of the over-represented closed conformations\\n (Oleinikovas et al., 2016) (Fig.', '4b).', 'Along with key residue R244 on\\n the opposite wall of the pocket, the opening of a-helices H11 and\\n H12 and key residues L220 and N276 dictate pocket opening and\\n allow two inhibitors to be accommodated (Horn and Shoichet,\\n 2004), while the three mentioned residues form a triad when the\\n pocket is closed.', 'In addition to the opening of the two helices, visual\\n inspection of the simulations indicated a deepening of the pocket.', 'As\\n a result, we chose features tracking the distance between the Ca of\\n residues L220 and N276 as well as that of their sidechains to monitor pocket opening as well as the distance between the Ca of I263\\n and I279 as a measure of pocket depth (Fig.', '4a).', 'The original study\\n used fpocket (Le Guilloux et al., 2009) to monitor pocket exposure\\n in each replica (Fig.', '4b, top).', 'The same was done on the clusters\\n obtained by CLoNe (Fig.', '4b, bottom), showing different levels of\\n pocket openness.', 'Corresponding cluster centers highlight key structural differences between each state (Fig.', '4c), which are representative of the feature distribution per cluster (Fig.', '4d, top).', 'Cluster\\n assignation follows the observation of the original publication,\\n where open states were more prevalent in the replica of medium\\n scaling (Fig.', '4d, bottom).', '3.3 Identifying dominant conformational motifs in\\n protein oligomerization\\n Another challenge in structural biology is the understanding of how\\n biomolecules oligomerize to distinctive functional states.', 'One of\\n these cases, the transmembrane a-helix of the Amyloid Precursor\\n Protein (termed APP hereafter), has recently been studied by our lab\\n through the high-throughput MD protocol DAFT (Docking Assay\\n For Transmembrane components, (Wassenaar et al., 2015)) in order\\n to identify which of two dimerization motifs is promoted depending\\n on the lipid composition of the synaptic plasma membrane\\n (Audagnotto et al., 2016).', 'The G700G704G708 motif is thought to\\n direct the binding of APP to regulators promoting cholesterol biosynthesis, while the G709A713 motif would bind to cholesterol molecules (Fig.', '5a).', 'Extracting features from molecular datasets is not\\n always straightforward.', 'Macromolecular movements possess an inherent redundancy due to the sheer number of internal degrees of\\n freedom or prior knowledge may be lacking in order to select meaningful features, such as those highlighted in Figures 3 and 4.', 'The use\\n of dimensionality reduction methods, such as principal component\\n analysis (PCA) has been seen for clustering of MD simulations\\n (Wolf and Kirschner, 2013) and can help identifying coordinates of\\n significance while discarding less useful dimensions.', 'The DAFT simulations of APP from (Audagnotto et al., 2016) are over 2 ms in total\\n and contain countless states, many corresponding to unbound\\n monomers.', 'The first principal component based on the Cartesian\\n coordinates of the coarse grain backbone covers 77% of the variability in the simulation, highlighting two clusters (Fig.', '5b).', 'The\\n blue cluster of lower amplitude corresponds to all states exhibiting\\n unbound monomers (Fig.', '5e, left), while the second cluster regroups\\n all the dimerized states regardless of motif.', 'Focusing on that cluster,\\n we calculated the pair-wise distances between the backbone atoms\\n of each motifs in both helices (Fig.', '5a) and reduced these features to\\n a two-dimensional principal space covering 94% of the variability\\n before clustering (Fig.', '5c).', 'We want to highlight CLoNe’s ability to\\n analyze the neighborhood of low-density clusters without influence\\n from high-density regions (Fig.', '5c and Supplementary Table S2).', 'Clusters in blue in Figure 5d all depict states close to the\\n G700G704G708 motifs and those in green the G709A713 motif.', 'In the\\n middle are two clusters, shown in brown, that we interpret as hybrid.', 'In all cases, the darker-shaded clusters of each group correspond to the closest to the optimal motif arrangement, while the\\n others can be considered as closely related metastable states\\n (Supplementary Fig.', 'S6).', 'Similar to the original study (Audagnotto\\n Fig.', '3.', 'Utilizing Gaussian cluster properties to extract centers as key biological states of the COQ9 membrane protein.', '(a) COQ9 and its associated features, which include an\\n internal angle h and its distance to the membrane d (left).', 'Cluster centers are shown on the right side of the panel.', '(b) Every frame is plotted in the mentioned feature space and\\n color coded according to their core cardinality (left) and cluster assignation (right), which follows the same color code as in (a).', 'Outliers are shown as black crosses and centers\\n as black stars\\n CLoNe: automated clustering of biomolecular structural ensembles 5\\n Downloaded from https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa742/5895303 by guest on 19 March 2021\\n et al., 2016), CLoNe finds the preferred dimerization motif to be\\n G700G704G708 as evidenced by the corresponding centers’ local densities, cluster population and core cardinality (Fig.', '5c and\\n Supplementary Table S2).', 'A comparison between CLoNe and other\\n state-of-the-art algorithms on all the structural data shown in this\\n section demonstrates the advantages of CLoNe for analysis of molecular structural ensembles (Supplementary Fig.', 'S7).', '4 Discussion\\n Many clustering methods rely on parameters that are often nontrivial to optimize or on random initial conditions that may drastically change the outcome.', 'Commonly used algorithms are generally\\n restricted to specific cluster properties, forcing the researcher\\n through a process of trial and error.', 'Moreover, choosing relevant\\n features from structural datasets is challenging and different features\\n may generate different cluster topologies, sometimes irrelevant.', 'CLoNe was designed with these issues in mind and aims to provide\\n a stream-lined analytic process to yield results rapidly along with\\n helpful visualization scripts to analyze and confirm the relevance of\\n the clusters in the target biological context (Supplementary Fig.', 'S1).', 'CLoNe’s only parameter regulates the size of the local neighborhood\\n considered around each data point, which can be regarded as cluster\\n sizing parameter.', 'Its value need only be decreased if clusters seem\\n too inclusive and vice-versa, making CLoNe an intuitive algorithm\\n to use in addition to its general applicability.', 'For structural datasets,\\n CLoNe is able to extract clusters as separate trajectories and provides scripts for their automatic loading in the visualization software\\n VMD (Humphrey et al., 1996).', 'For larger macromolecules, the concept of a conformational state is blurry, hard to determine and often\\n depends on context.', 'It is not always clear which features to use to\\n obtain an accurate partition of the structural ensemble.', 'The results\\n obtained on COQ9 can be obtained hypothesis-free on raw spatial\\n coordinates or using PCA to extract relevant features\\n (Supplementary Fig.', 'S4).', 'This was done in the case of APP as well as\\n to reduce an otherwise redundant feature space to one of lower\\n dimensionality.', 'If one wishes to disentangle internal from overall\\n motion, dihedral PCA was used with success to study peptide folding (Altis et al., 2007; Mu et al., 2004).', 'However, when other\\n features than the chosen ones can be expected to exhibit motions of\\n larger amplitudes, PCA will favor the latter over the former.', 'This is\\n true for the TEM1 b-lactamase, where internal structural motions\\n will be more prevalent than the fluctuations of the selected key\\n pocket residues.', 'In such cases, a feature-based approach is to be preferred.', 'Alternatively, some will advocate the use of time-lagged independent component analysis (TICA) (Naritomi and Fuchigami,\\n 2011) instead.', 'TICA was found to be the better alternative for building Markov state models (Husic and Pande, 2018; Pe´rez-Herna´ndez\\n et al., 2013).', 'However, in cases where large amplitude fluctuations\\n are the target or when there is redundancy in features, we believe\\n that PCA remains a safe approach.', 'As the conformational ensembles presented in this study tend to\\n exhibit Gaussian distributions, CLoNe may thus be used to extract\\n cluster centers as higher probability states.', 'Such states offer an overview of the ensemble and may serve as starting models for building\\n Markov state models in general.', 'Moreover, the precision of the classification achieved by CLoNe enables the identification of dominant\\n biological states from large datasets.', 'Beyond the case of APP,\\n CLoNe identified different key pocket conformations in the case of\\n TEM1 b-lactamase.', 'Further clustering efforts on this system should\\n target the different positions of R244, which was not tracked in this\\n study but was previously shown to play a dual role between TEM1’s\\n active site and allosteric pocket (Horn and Shoichet, 2004).', 'CLoNe\\n may then be used as a pre-processing tool prior to small-molecule\\n docking studies, where accounting for receptor flexibility is an active\\n development area (Kokh et al., 2011; Vahl Quevedo et al., 2014).', 'Integrative modeling aims at incorporating data from multiple\\n sources to determine the structure of macromolecular complexes.', 'Such hybrid strategies typically combine low resolution data of\\n whole complexes with high resolution structures of their components so as to predict the quaternary structure of the former\\n (Cassidy et al., 2018).', 'This process is however severely hindered by\\n structural dynamics differing between a complex and its isolated\\n components.', 'For this reason, many hybrid modeling strategies now\\n incorporate some form of dynamics to bridge this gap (Malhotra\\n et al., 2019; Tamo` et al., 2015).', 'While we previously utilized classical MD for the prediction of heptameric aerolysin pores (Degiacomi\\n et al., 2013; Degiacomi and Dal Peraro, 2013), such an approach\\n would not be feasible for heteromultimeric assemblies where\\n Fig.', '4.', 'Identification of different opening states of the allosteric cryptic pocket in TEM1 b-lactamase.', '(a) apo and Holo structures (left and right, respectively).', 'Allosteric inhibitors are shown in gray and white.', 'Features following helical opening include the distance between Ca atoms of N276 and L220 (medium blue) and the Cc of their sidechain\\n (light blue).', 'Pocket depth is monitored by the distance between Ca-carbons of I263 and 279 (dark blue).', '(b) The pocket exposure calculated using the fpocket software for the\\n original replicas (top) and for each clusters (bottom).', 'The dotted line in both is the reference value of the holo crystal structure used in the original paper.', '(c) The center of each\\n cluster in cartoon representation on top of a surface representation of the allosteric pocket, highlighting the different states of helical openness and pocket depth.', 'The triad\\n N276-L220-R244 governing pocket opening and closing are shown as gray sticks.', '(c) The distribution of each feature for each cluster (top) and the cluster assignation along\\n the three chosen replicas (bottom)\\n 6 S.Tra¨ ger et al.', 'Downloaded from https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa742/5895303 by guest on 19 March 2021\\n multiple conformational ensembles are required simultaneously.', 'Reducing them to their crucial components may enable the structural characterization of large macromolecular complexes, which\\n may otherwise be intractable.', 'Applied to the fields of smallmolecule docking, integrative modeling and structural dynamics\\n studies, CLoNe presents itself as a versatile and powerful tool for\\n modern computational structural biology.', 'Acknowledgements\\n We thank Vladimiras Oleinikovas and Francesco L. Gervasio for providing\\n the simulations of the TEM1 b-lactamase as well as offering general advice;\\n Lucien F. Krapp and Romain Groux for helpful discussions.', 'Funding\\n M.D.P.', 'lab was supported by the Swiss National Science Foundation (grants\\n number 200021_157217 and 31003A_170154).', 'Conflict of Interest: none declared.', 'References\\n Abriata,L.A.', 'and Dal Peraro,M.', '(2020) Will cryo-electron microscopy shift\\n the current paradigm in protein structure prediction?', 'J. Chem.', 'Inf.', 'Model.,\\n 60, 2443–2447.', 'Altis,A.', 'et al.', '(2007) Dihedral angle principal component analysis of molecular\\n dynamics simulations.', 'J. Chem.', 'Phys., 126, 244111.', 'Amaro,R.E.', 'et al.', '(2018) Ensemble docking in drug discovery.', 'Biophys.', 'J.,\\n 114, 2271–2278.', 'Ankerst,M.', 'et al.', '(1999) OPTICS: Ordering Points to Identify the Clustering\\n Structure.', 'In: Proceedings of the 1999 ACM SIGMOD International\\n Conference on Management of Data, SIGMOD ’99, pp.', '49–60.', 'ACM, New\\n York, NY, USA.', 'Audagnotto,M.', 'and Dal Peraro,M.', '(2017) Protein post-translational modifications: in silico prediction tools and molecular modeling.', 'Comput.', 'Struct.', 'Biotechnol.', 'J., 15, 307–319.', 'Audagnotto,M.', 'et al.', '(2016) Effect of the synaptic plasma membrane on the\\n stability of the amyloid precursor protein homodimer.', 'J. Phys.', 'Chem.', 'Lett.,\\n 7, 3572–3578.', 'Barducci,A.', 'et al.', '(2011) Metadynamics.', 'Wiley Interdiscip.', 'Rev.', 'Comput.', 'Mol.', 'Sci., 1, 826–843.', 'Beauchamp,K.A.', 'et al.', '(2012) Simple few-state models reveal hidden complexity in protein folding.', 'Proc.', 'Natl.', 'Acad.', 'Sci.', 'USA, 109, 17807–17813.', 'Bhattacharyya,A.', '(1943) On a measure of divergence between two statistical\\n populations defined by their probability distributions.', 'Bull.', 'Calcutta Math.', 'Soc., 35, 99–109.', 'Bussi,G.', '(2014) Hamiltonian replica exchange in GROMACS: a flexible implementation.', 'Mol.', 'Phys., 112, 379–384.', 'Cassidy,C.K.', 'et al.', '(2018) CryoEM-based hybrid modeling approaches for\\n structure determination.', 'Curr.', 'Opin.', 'Microbiol., 43, 14–23.', 'Chang,H.', 'and Yeung,D.-Y.', '(2008) Robust path-based spectral clustering.', 'Pattern Recogn., 41, 191–203.', 'Chavent,M.', 'et al.', '(2016) Molecular dynamics simulations of membrane proteins and their interactions: from nanoscale to mesoscale.', 'Curr.', 'Opin.', 'Struct.', 'Biol., 40, 8–16.', 'Cheng,L.S.', 'et al.', '(2008) Ensemble-based virtual screening reveals potential\\n novel antiviral compounds for avian influenza neuraminidase.', 'J. Med.', 'Chem., 51, 3878–3894.', 'd’Errico,M.', 'et al.', '(2018) Automatic topography of high-dimensional data sets\\n by non-parametric density peak clustering.', 'arXiv:1802.10549v1 [stat.ML]\\n De Paris,R.', 'et al.', '(2015) Clustering molecular dynamics trajectories for optimizing docking experiments.', 'Comput.', 'Intell.', 'Neurosci., 2015, 1–9.', 'de Souza,V.C.', 'et al.', '(2017) Clustering algorithms applied on analysis of protein molecular dynamics.', 'In: 2017 IEEE Latin American Conference on\\n Computational Intelligence (LA-CCI).', 'Arequipa, 2017, pp.', '1–6.', 'https://doi.', 'org/10.1109/LA-CCI.2017.8285695\\n De Vivo,M.', 'et al.', '(2016) Role of molecular dynamics and related methods in\\n drug discovery.', 'J. Med.', 'Chem., 59, 4035–4061.', 'Degiacomi,M.T.', 'and Dal Peraro,M.', '(2013) Macromolecular symmetric assembly prediction using swarm intelligence dynamic modeling.', 'Structure,\\n 21, 1097–1106.', 'Degiacomi,M.T.', 'et al.', '(2013) Molecular assembly of the aerolysin pore reveals\\n a swirling membrane-insertion mechanism.', 'Nat.', 'Chem.', 'Biol., 9, 623–629.', 'Doerr,S.', 'et al.', '(2016) HTMD: high-throughput molecular dynamics for molecular discovery.', 'J. Chem.', 'Theory Comput., 12, 1845–1852.', 'Du,M.', 'et al.', '(2016) Study on density peaks clustering based on k-nearest neighbors and principal component analysis.', 'Knowl.', 'Based Syst., 99, 135–145.', 'Ester,M.', 'et al.', '(1996) A density-based algorithm for discovering clusters a\\n density-based algorithm for discovering clusters in large spatial databases\\n with noise.', 'In: Proceedings of the Second International Conference on\\n Knowledge Discovery and Data Mining, KDD’96, pp.', '226–231.', 'AAAI\\n Press, Portland, OR, USA.', 'Fonti,G.', 'et al.', '(2019) KAP1 is an antiparallel dimer with a natively functional\\n asymmetry.', 'Life Science Alliance, 2(4), e201900349.', 'Frank,J.', '(2018) New opportunities created by single-particle cryo-EM: the\\n mapping of conformational space.', 'Biochemistry, 57, 888–888.', 'Fra¨nti,P., and Sieranoja,S.', '(2018) K-means properties on six clustering benchmark datasets.', 'Applied Intelligence, 48, 4743–4759.', '10.1007/s10489-\\n 018-1238-7\\n Fig.', '5.', 'Classification and frequency estimation of APP dimerization motifs in the\\n plasma membrane.', '(a) The two known dimerization motifs of the transmembrane\\n helix of APP.', 'Membrane is depicted with gray spheres, the G700G704G708 motif\\n with teal spheres and the G709A713 motif with gold spheres.', '(b) The black line\\n depicts the probability density distribution of the trajectory along the first principal\\n component, which covers 77% of variability.', 'Obtained clusters are shown in colors\\n underneath, with their centers as black stars.', 'Below the clusters is the distribution of\\n core cardinalities, with the corresponding color bar below it.', '(c) The cluster of\\n bound dimers of (b) was extracted and the pair-wise distances between the backbone atoms of each motifs in both helices was computed for every frame.', 'The plot\\n shows the first two principal components of the resulting dataset (with eigenvalues\\n of 0.66 and 0.28, respectively), where each point is color-coded according to the\\n cardinality of its core.', 'Stars represent cluster centers.', '(d) The clusters obtained from\\n the aforementioned dataset of bound conformations.', '(e) Renders of the unbound\\n and dimerized APP cluster centers and their respective frequency after outlier removal.', 'Dimerized APP centers shown correspond to the dark blue, dark brown and\\n dark green clusters.', 'Sidechains have been hidden for better visualization\\n CLoNe: automated clustering of biomolecular structural ensembles 7\\n Downloaded from https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa742/5895303 by guest on 19 March 2021\\n Fu,L.', 'and Medico,E.', '(2007) FLAME, a novel fuzzy clustering method for the\\n analysis of DNA microarray data.', 'BMC Bioinformatics, 8, 3.', 'Gionis,A.', 'et al.', '(2007) Clustering aggregation.', 'ACM Trans.', 'Knowl.', 'Discov.', 'Data, 1, 4.', 'Hamelberg,D.', 'et al.', '(2004) Accelerated molecular dynamics: a promising and\\n efficient simulation method for biomolecules.', 'J. Chem.', 'Phys., 120,\\n 11919–11929.', 'Horn,J.R.', 'and Shoichet,B.K.', '(2004) Allosteric inhibition through core disruption.', 'J. Mol.', 'Biol, 336, 1283–1291.', 'Humphrey,W.', 'et al.', '(1996) VMD: visual molecular dynamics.', 'J. Mol.', 'Graph,\\n 14, 33–38.', 'Husic,B.E.', 'and Pande,V.S.', '(2018) Markov state models: from an art to a science.', 'J.', 'Am.', 'Chem.', 'Soc., 140, 2386–2396.', 'Husic,B.E.', 'and Pande,V.S.', '(2017) Ward clustering improves cross-validated Markov\\n state models of protein folding.J.', 'Chem.', 'Theory Comput., 13, 963–967.', 'Jain,A.K.', '(2010) Data clustering: 50 years beyond K-means.', 'Pattern Recognit.', 'Lett., 31, 651–666.', 'Jones,E.', 'et al.', '(2001) SciPy: open source scientific tools for Python.', '10.1038/s41592-019-0686-2.', 'Kokh,D.B.', 'et al.', '(2011) Receptor flexibility in small-molecule docking calculations.', 'Wiley Interdiscip.', 'Rev.', 'Comput.', 'Mol.', 'Sci., 1, 298–314.', 'Le Guilloux,V.', 'et al.', '(2009) Fpocket: an open source platform for ligand\\n pocket detection.', 'BMC Bioinformatics, 10, 168.', 'Liang,Z.', 'and Chen,P.', '(2016) Delta-density based clustering with a\\n divide-and-conquer strategy: 3DC clustering.', 'Pattern Recognit.', 'Lett., 73, 52–59.', 'Lohman,D.C.', 'et al.', '(2019) An isoprene lipid-binding protein promotes eukaryotic coenzyme Q biosynthesis.', 'Mol.', 'Cell, 73, 763–774.e10.', 'Lohman,D.C.', 'et al.', '(2014) Mitochondrial COQ9 is a lipid-binding protein\\n that associates with COQ7 to enable coenzyme Q biosynthesis.', 'Proc.', 'Natl.', 'Acad.', 'Sci.', 'USA, 111, E4697–E4705.', 'Malhotra,S.', 'et al.', '(2019) Modelling structures in cryo-EM maps.', 'Curr.', 'Opin.', 'Struct.', 'Biol., 58, 105–114.', 'McGibbon,R.T.', 'et al.', '(2015) MDTraj: a modern open library for the analysis\\n of molecular dynamics trajectories.', 'Biophys.', 'J., 109, 1528–1532.', 'McKiernan,K.', 'A. et al.. (2017) Modeling the mechanism of CLN025 beta--\\n hairpin formation.', 'The Journal of Chemical Physics, 147, 104107.', 'Mehmood,R.', 'et al.', '(2016) Clustering by fast search and find of density peaks\\n via heat diffusion.', 'Neurocomputing, 208, 210–217.', 'Mu,Y.', 'et al.', '(2004) Energy landscape of a small peptide revealed by dihedral\\n angle principal component analysis.', 'Proteins Struct.', 'Funct.', 'Bioinform., 58,\\n 45–52.', 'Naritomi,Y.', 'and Fuchigami,S.', '(2011) Slow dynamics in protein fluctuations\\n revealed by time-structure based independent component analysis: the case\\n of domain motions.', 'J. Chem.', 'Phys., 134, 065101.', 'Noe´,F.', 'et al.', '(2019) Boltzmann generators: sampling equilibrium states of\\n many-body systems with deep learning.', 'Science, 365, eaaw1147.', 'Oleinikovas,V.', 'et al.', '(2016) Understanding cryptic pocket formation in protein targets by enhanced sampling simulations.', 'J.', 'Am.', 'Chem.', 'Soc., 138,\\n 14257–14263.', 'Paris,R.D.', 'et al.', '(2015) An effective approach for clustering InhA molecular\\n dynamics trajectory using substrate-binding cavity features.', 'PLoS One, 10,\\n e0133172.', 'Pedregosa,F.', 'et al.', '(2011) Scikit-learn: machine learning in Python.', 'J. Mach.', 'Learn.', 'Res., 12, 2825–2830.', 'Peng,J.', 'et al.', '(2018) Clustering algorithms to analyze molecular dynamics\\n simulation trajectories for complex chemical and biological systems.', 'Chin.', 'J. Chem.', 'Phys., 31, 404–420.', 'Pe´rez-Herna´ndez,G.', 'et al.', '(2013) Identification of slow molecular order\\n parameters for Markov model construction.', 'J. Chem.', 'Phys., 139, 015102.', 'Rodriguez,A.', 'et al.', '(2018) Computing the free energy without collective variables.', 'J. Chem.', 'Theory Comput., 14, 1206–1215.', 'Rodriguez,A.', 'and Laio,A.', '(2014) Clustering by fast search and find of density\\n peaks.', 'Science, 344, 1492–1496.', 'Salmaso,V., and Moro,S.', '(2018) Bridging molecular docking to molecular dynamics in exploring ligand-protein recognition process: an overview.', 'Front.', 'Pharmacol, 9, 923.', 'Seabold,S.', 'and Perktold,J.', '(2010) Statsmodels: econometric and statistical\\n modeling with python.', 'In: 9th Python in Science Conference.', 'https://www.', 'statsmodels.org/devel/#citation.', 'Shao,J.', 'et al.', '(2007) Clustering molecular dynamics trajectories: 1.', 'Characterizing the performance of different clustering algorithms.', 'J. Chem.', 'Theory Comput., 3, 2312–2334.', 'Shirts,M.', 'and Pande,V.S.', '(2000) Screen savers of the World Unite!', 'Science,\\n 290, 1903–1904.', 'Sultan,M.M.', 'et al.', '(2018) Transferable neural networks for enhanced sampling of protein dynamics.', 'J. Chem.', 'Theory Comput., 14, 1887–1894.', 'Tamo`,G.E.', 'et al.', '(2015) The importance of dynamics in integrative modeling\\n of supramolecular assemblies.', 'Curr.', 'Opin.', 'Struct.', 'Biol., 31, 28–34.', 'Vahl Quevedo,C.', 'et al.', '(2014) A strategic solution to optimize molecular\\n docking simulations using Fully-Flexible Receptor models.', 'Expert Syst.', 'Appl., 41, 7608–7620.', 'Wang,W.', 'et al.', '(2018) Constructing Markov state models to elucidate the\\n functional conformational changes of complex biomolecules.', 'Wiley\\n Interdiscip.', 'Rev.', 'Comput.', 'Mol.', 'Sci., 8, e1343.', 'Wang,X.-F. and Xu,Y.', '(2017) Fast clustering using adaptive density peak detection.', 'Stat.', 'Methods Med.', 'Res., 26, 2800–2811.', 'Ward,J.H.', '(1963) Hierarchical grouping to optimize an objective function.', 'J.', 'Am.', 'Stat.', 'Assoc., 58, 236–244.', 'Wassenaar,T.A.', 'et al.', '(2015) High-throughput simulations of dimer and trimer assembly of membrane proteins.', 'the DAFT Approach.', 'J. Chem.', 'Theory\\n Comput., 11, 2278–2291.', 'Wolf,A.', 'and Kirschner,K.N.', '(2013) Principal component and clustering analysis on molecular dynamics data of the ribosomal L11\\x0523S subdomain.', 'J.\\n Mol.', 'Model., 19, 539–549.', 'Xie,J.', 'et al.', '(2016) Robust clustering by detecting density peaks and assigning\\n points based on fuzzy weighted K-nearest neighbors.', 'Inf.', 'Sci., 354, 19–40.', 'Zhang,W.', 'and Li,J.', '(2015) Extended fast search clustering algorithm: widely\\n density clusters, no density peaks.', 'arXiv:1505.05610 [cs.DS].', '8 S.Tra¨ ger et al.', 'Downloaded from https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btaa742/5895303 by guest on 19 March 2021']\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        },
        "id": "BzSxRz_YDJ_B",
        "outputId": "4236b963-dcd8-42ad-ed81-e43dd4c88089"
      },
      "source": [
        "lsa_summary[0]"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            },
            "text/plain": [
              "'['"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 35
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nySyffmVAISY",
        "outputId": "39f19e70-0bcc-4adb-b252-075d29ded1f0"
      },
      "source": [
        "myquery= input(\"Please enter the information to be searched \")\n",
        "data =google_search(API_KEY,SEARCH_ENGINE_ID,myquery)\n",
        "results(data)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Please enter the information to be searched Structural bioinformatics\\\\n CLoNe: automated clustering based on local density\\\\n neighborhoods for application to biomolecular structural\\\\n ensembles\\\\n Sylvain Tra¨ ger1,2, Giorgio Tamo` 1,2, Deniz Aydin1,2, Giulia Fonti1,2,\\\\n Martina Audagnotto1,2 and Matteo Dal Peraro1,2,*\\\\n 1\\\\n Institute of Bioengineering, School of Life Sciences, Ecole Polytechnique Fe´ de´ rale de Lausanne, Lausanne 1025, Switzerland and 2\\\\n Swiss Institute of Bioinformatics, Lausanne 1015, Switzerland\\\\n *To whom correspondence should be addressed.', 'Associate Editor: Lenore Cowen\\\\n Received on January 8, 2020; revised on July 14, 2020; editorial decision on August 12, 2020; accepted on August 18, 2020\\\\n Abstract\\\\n Motivation: Proteins are intrinsically dynamic entities.', 'Flexibility sampling methods, such as molecular dynamics or\\\\n those arising from integrative modeling strategies, are now commonplace and enable the study of molecular conformational landscapes in many contexts\n",
            "Title: CLoNe: automated clustering based on local density neighborhoods ...\n",
            "Description: Aug 21, 2020 ... neighborhoods for application to biomolecular structural ensembles. Sylvain Tra¨ \n",
            "ger1,2, Giorgio Tamo`1,2, Deniz Aydin1,2, Giulia Fonti1,2,.\n",
            "URL: https://academic.oup.com/bioinformatics/advance-article-pdf/doi/10.1093/bioinformatics/btaa742/36453399/btaa742.pdf \n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rfA1A9jNEqk0",
        "outputId": "1fbbc5c8-8ee4-4437-d1a6-6ba69726512b"
      },
      "source": [
        "url = \"https://academic.oup.com/bioinformatics/advance-article-pdf/doi/10.1093/bioinformatics/btaa742/36453399/btaa742.pdf  \"\n",
        "\n",
        "# Requests URL and get response object \n",
        "response = requests.get(url) \n",
        "i=1\n",
        "pdf = open(\"pdf\"+str(i)+\".pdf\", 'wb') \n",
        "pdf.write(response.content) \n",
        "pdf.close() \n",
        "print(\"File \", i, \" downloaded\")"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "File  1  downloaded\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 369
        },
        "id": "EuPPfLYDFHqY",
        "outputId": "247beca3-4927-4bf1-9184-3e16a8591d76"
      },
      "source": [
        "import PyPDF2\n",
        "from PyPDF2 import PdfFileReader\n",
        "  \n",
        "# creating a pdf file object\n",
        "pdfFileObj = open('pdf1.pdf', 'rb')\n",
        "  \n",
        "# creating a pdf reader object\n",
        "pdfReader = PyPDF2.PdfFileReader(pdfFileObj)\n",
        "with open(\"pdf1.pdf\", 'rb') as f:\n",
        "  pdf = PdfFileReader(f)\n",
        "  information = pdf.getDocumentInfo()\n",
        "  number_of_pages = pdf.getNumPages()\n",
        "\n",
        "txt = f\"\"\"\n",
        "Information about {\"pdf1.pdf\"}: \n",
        "Author: {information.author}\n",
        "Creator: {information.creator}\n",
        "Producer: {information.producer}\n",
        "Subject: {information.subject}\n",
        "Title: {information.title}\n",
        "Number of pages: {number_of_pages}\n",
        "    \"\"\"\n",
        "\n",
        "print(txt)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "PdfReadError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mPdfReadError\u001b[0m                              Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-38-27542030cf8b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;31m# creating a pdf reader object\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m \u001b[0mpdfReader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPyPDF2\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPdfFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpdfFileObj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"pdf1.pdf\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'rb'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m   \u001b[0mpdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPdfFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/PyPDF2/pdf.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, stream, strict, warndest, overwriteWarnings)\u001b[0m\n\u001b[1;32m   1082\u001b[0m             \u001b[0mstream\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mBytesIO\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb_\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfileobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1083\u001b[0m             \u001b[0mfileobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1084\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstream\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1085\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstream\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstream\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1086\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/PyPDF2/pdf.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, stream)\u001b[0m\n\u001b[1;32m   1695\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mstream\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtell\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0mlast1K\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1696\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mutils\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPdfReadError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"EOF marker not found\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1697\u001b[0;31m             \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadNextEndLine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstream\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1698\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mdebug\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"  line:\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1699\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/PyPDF2/pdf.py\u001b[0m in \u001b[0;36mreadNextEndLine\u001b[0;34m(self, stream)\u001b[0m\n\u001b[1;32m   1935\u001b[0m             \u001b[0;31m# Prevent infinite loops in malformed PDFs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1936\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mstream\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtell\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1937\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mutils\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPdfReadError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Could not read malformed PDF file\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1938\u001b[0m             \u001b[0mx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstream\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1939\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mdebug\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"  x:\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"%x\"\u001b[0m\u001b[0;34m%\u001b[0m\u001b[0mord\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mPdfReadError\u001b[0m: Could not read malformed PDF file"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 420
        },
        "id": "6dDQhvKTAey_",
        "outputId": "31a6809d-3123-4b81-befb-e73504336b58"
      },
      "source": [
        "import newspaper \n",
        "URL_link =only_link(data)\n",
        "all_scraped_documents =[]  \n",
        "i= 0\n",
        "# Parse through each url and display its content \n",
        "for url in URL_link: \n",
        "    print(\"First Run\")\n",
        "    url_i = newspaper.Article(url=\"%s\" % (url), language='en') \n",
        "    url_i.download() \n",
        "    url_i.parse() \n",
        "    all_scraped_documents.append(url_i.text)\n",
        "    i = i+1\n",
        "    print(url_i.text)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "First Run\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "error",
          "ename": "ArticleException",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mArticleException\u001b[0m                          Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-36-536ddf7c9619>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      8\u001b[0m     \u001b[0murl_i\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnewspaper\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mArticle\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"%s\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlanguage\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'en'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m     \u001b[0murl_i\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdownload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m     \u001b[0murl_i\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     11\u001b[0m     \u001b[0mall_scraped_documents\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl_i\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtext\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m     \u001b[0mi\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/newspaper/article.py\u001b[0m in \u001b[0;36mparse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    189\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    190\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 191\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mthrow_if_not_downloaded_verbose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    192\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    193\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdoc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_parser\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfromstring\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhtml\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/newspaper/article.py\u001b[0m in \u001b[0;36mthrow_if_not_downloaded_verbose\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    530\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdownload_state\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mArticleDownloadState\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mFAILED_RESPONSE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    531\u001b[0m             raise ArticleException('Article `download()` failed with %s on URL %s' %\n\u001b[0;32m--> 532\u001b[0;31m                   (self.download_exception_msg, self.url))\n\u001b[0m\u001b[1;32m    533\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    534\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mthrow_if_not_parsed_verbose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mArticleException\u001b[0m: Article `download()` failed with 403 Client Error: Forbidden for url: https://academic.oup.com/bioinformatics/advance-article-pdf/doi/10.1093/bioinformatics/btaa742/36453399/btaa742.pdf on URL https://academic.oup.com/bioinformatics/advance-article-pdf/doi/10.1093/bioinformatics/btaa742/36453399/btaa742.pdf"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "7YaOe4MTBuHZ"
      },
      "source": [
        "sentencesweb=sent_tokenize(textextract2)\n",
        "text_embeddings = model.encode(sentencesweb, batch_size = 8, show_progress_bar = True)\n",
        "text_embeddings2 = model.encode(sentencesAtten, batch_size = 8, show_progress_bar = True)\n",
        "np.shape(text_embeddings)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "HrHuQ__fB4VJ"
      },
      "source": [
        "similarities = cosine_similarity(text_embeddings,text_embeddings2)\n",
        "similarities_sorted = similarities.argsort()\n",
        "# Get list of similarity indices i.e. doc at index 0 simialr with doc at index 1169 below.\n",
        "id_1 = []\n",
        "id_2 = []\n",
        "score = []\n",
        "for index,array in enumerate(similarities_sorted):\n",
        "    id_1.append(index)\n",
        "    id_2.append(array[-2])\n",
        "    score.append(similarities[index][array[-2]])"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "FQgy4rwBB-vz"
      },
      "source": [
        "sum_of_sims =(np.sum(index_df[\"score\"]))\n",
        "print(sum_of_sims)\n",
        "percentage_of_similarity = round(float((sum_of_sims / len(sentencesweb)) * 100))\n",
        "print(f'Average similarity float: {float(sum_of_sims / len(sentencesweb))}')\n",
        "print(f'Average similarity percentage: {float(sum_of_sims / len(sentencesweb)) * 100}')\n",
        "print(f'Average similarity rounded percentage: {percentage_of_similarity}')"
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}